---
title: "Medical Intensive Care Unit Study"
author: "Nicholas P Dylla, Huaiying Lin"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: yeti
    highlight: pygments
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
editor_options: 
  chunk_output_type: console
header-includes:
  - |
    ```{=latex}
    \usepackage{fvextra}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{
      showspaces = false,
      showtabs = false,
      breaklines,
      commandchars=\\\{\}
---

# {.tabset}
```{r, include=FALSE}
knitr::opts_chunk$set(
  warning = FALSE, message = FALSE, error = FALSE, tidy = "styler", tidy.opts = list(width.cutoff = 60),
  fig.height = 10, fig.width = 16, fig.paths = "./Results/"
)
```

## Load Packages
```{r, load-packages}
# Install `remotes`
if (!require("remotes", quietly = TRUE)) {
  install.packages("remotes")
}

# Install devtools
if (!require("devtools", quietly = TRUE)) {
  install.packages("devtools")
}

# Install `librarian`
if (!require("librarian", quietly = TRUE)) {
  remotes::install_github("DesiQuintans/librarian")
}

# Install `splitTools`
if (!require("splitTools", quietly = TRUE)) {
  remotes::install_github("mayer79/splitTools")
}

# Install `BiocManager`
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

# Install `Biobase` to enable `librarian` to install missing packages
if (!require("Biobase", quietly = TRUE)) {
  BiocManager::install("Biobase")
}


# Load/Install Packages
shelf(
  sckott / cowsay,
  ggpubr,
  tidyverse,
  tableone,
  glmnet,
  rstatix,
  reticulate,
  caret,
  survival,
  survminer,
  cutpointr,
  shadowtext,
  pROC,
  gtsummary,
  gt,
  Maaslin2,
  vegan,
  pairwiseAdonis,
  umap,
  factoextra,
  EnhancedVolcano,
  ggpirate,
  splitTools,
  yingtools2,
  patchwork,
  janitor,
  labelled,
  conflicted,
  survRM2
)

# Package conflicts
{
  conflict_prefer("select", "dplyr")
  conflict_prefer("mutate", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("slice", "dplyr")
  conflict_prefer("between", "dplyr")
  conflict_prefer("annotate", "ggplot2")
  conflict_prefer("cummax", "yingtools2")
  conflict_prefer("simplify", "igraph")
  conflict_prefer("predict", "stats")
  conflict_prefer("splsda", "mixOmics")
  conflict_prefer("cbind", "base")
  conflict_prefer("make.names", "base")
  conflict_prefer("unique", "base")
  conflict_prefer("as.data.frame", "base")
  conflict_prefer("setdiff", "base")
  conflict_prefer("cutpoints", "cutpointr")
  conflict_prefer("chisq.test", "stats")
}


# color palette
devtools::source_url("https://github.com/yingeddi2008/DFIutility/blob/master/getRdpPal.R?raw=TRUE")

cowsay::say(
  "Fun microbiome analysis is. Hmmm.",
  by = "yoda",
  what_color = "black",
  by_color = c("#2B7F29", "#6EBA35", "#2B7F29")
)


# ggplot theme shortcuts
et <- element_text
eb <- element_blank
er <- element_rect
el <- element_line

# Set path to python3
use_condaenv("/Users/nick/miniconda3/bin/python")
```

## Load Data Image
```{r, load-data-image, eval=FALSE}
# R image
load("./Data/MICU_Data_Anon.revision.RData")
```

## Load Data
```{r, load-data}
# Load in sample lookup list 
first_samp_list_anon <- readRDS("./Data/first_samp_list_anon.rds")

# Load in clinical variables
micu_new_anon <- readRDS("./Data/micu_new_anon.rds") # new micu

# Split data into stratified partitions
micu_index <-
  partition(
    y = first_samp_list_anon$thirtyday_mortality_overall,
    p = c(train = 0.75, test = 0.25),
    type = "stratified",
    seed = 543,
    shuffle = TRUE
  )

# Original cohort lookup
first_samp_list_oc <-
  first_samp_list_anon[micu_index$train, ]

# Validation cohort lookup
first_samp_list_vc <-
  first_samp_list_anon[micu_index$test, ]

# Original cohort clinical variables
micu_new_nocovid_oc <- micu_new_anon %>%
  right_join(first_samp_list_oc)

# Validation cohort clinical variables
micu_new_nocovid_vc <- micu_new_anon %>%
  right_join(first_samp_list_vc)

taxdmp <- readRDS("./Data/taxdmp.rds")

metaphlan <- readRDS("./Data/metaphlan.rds")

pal <- getRdpPal(metaphlan)

metab_quant_imp_tot_mM <- readRDS("./Data/metab_quant_imp_tot_mM.rds")

metab_qual_imp_tot <- readRDS("./Data/metab_qual_imp_tot.rds")

cri_rxmar_abx_long <- readRDS("./Data/cri_rxmar_abx_long.rds")
```

## Custom Functions
```{r, custom-functions}
# Need to increase Maaslin's color palette
maaslin2_association_plots <-
  function(metadata,
           features,
           output_results,
           write_to = "./",
           figures_folder = "./figures/",
           max_pngs = 10,
           save_scatter = FALSE) {
    if (is.character(metadata)) {
      metadata <- read.table(
        metadata,
        header = TRUE,
        row.names = 1,
        sep = "\t",
        fill = FALSE,
        comment.char = "",
        check.names = FALSE
      )
    }
    if (is.character(features)) {
      features <- read.table(
        features,
        header = TRUE,
        row.names = 1,
        sep = "\t",
        fill = FALSE,
        comment.char = "",
        check.names = FALSE
      )
    }
    common_rows <- intersect(rownames(features), rownames(metadata))
    input_df_all <- cbind(
      features[common_rows, , drop = FALSE],
      metadata[common_rows, , drop = FALSE]
    )
    if (is.character(output_results)) {
      output_df_all <- read.table(
        output_results,
        header = TRUE,
        row.names = NULL,
        sep = "\t",
        fill = FALSE,
        comment.char = "",
        check.names = FALSE
      )
    } else {
      output_df_all <- output_results
    }
    if (dim(output_df_all)[1] < 1) {
      print("There are no associations to plot!")
      return(NULL)
    }
    logging::loginfo(
      paste(
        "Plotting associations from most",
        "to least significant,",
        "grouped by metadata"
      )
    )
    metadata_types <- unlist(output_df_all[, "metadata"])
    metadata_labels <-
      unlist(metadata_types[!duplicated(metadata_types)])
    metadata_number <- 1
    saved_plots <- list()
    for (label in metadata_labels) {
      saved_plots[[label]] <- list()
      plot_file <- paste(write_to,
        "/",
        gsub(
          "[^[:alnum:]_]",
          "_", label
        ),
        ".pdf",
        sep = ""
      )
      data_index <- which(label == metadata_types)
      logging::loginfo(
        "Plotting data for metadata number %s, %s",
        metadata_number,
        label
      )
      pdf(plot_file,
        width = 2.65,
        height = 2.5,
        onefile = TRUE
      )
      x <- NULL
      y <- NULL
      count <- 1
      for (i in data_index) {
        x_label <- as.character(output_df_all[i, "metadata"])
        y_label <- as.character(output_df_all[i, "feature"])
        results_value <- as.character(output_df_all[i, "value"])
        qval <- as.numeric(output_df_all[i, "qval"])
        coef_val <- as.numeric(output_df_all[i, "coef"])
        input_df <- input_df_all[c(x_label, y_label)]
        colnames(input_df) <- c("x", "y")
        temp_plot <- NULL
        if (is.numeric(input_df[1, "x"]) &
          length(unique(input_df[["x"]])) >
            2) {
          logging::loginfo(
            "Creating scatter plot for continuous data, %s vs %s",
            x_label,
            y_label
          )
          temp_plot <- ggplot2::ggplot(
            data = input_df,
            ggplot2::aes(
              as.numeric(as.character(x)),
              as.numeric(as.character(y))
            )
          ) +
            ggplot2::geom_point(
              fill = "darkolivegreen4",
              color = "black",
              alpha = 0.5,
              shape = 21,
              size = 1,
              stroke = 0.15
            ) +
            ggplot2::scale_x_continuous(limits = c(
              min(input_df["x"]),
              max(input_df["x"])
            )) +
            ggplot2::scale_y_continuous(limits = c(
              min(input_df["y"]),
              max(input_df["y"])
            )) +
            ggplot2::stat_smooth(
              method = "glm",
              size = 0.5,
              color = "blue",
              na.rm = TRUE
            ) +
            ggplot2::guides(alpha = "none") +
            ggplot2::labs("") +
            ggplot2::xlab(x_label) +
            ggplot2::ylab(y_label) +
            nature_theme(input_df[, "x"], y_label) +
            ggplot2::annotate(
              geom = "text",
              x = Inf,
              y = Inf,
              hjust = 1,
              vjust = 1,
              label = sprintf(
                "FDR: %s\nCoefficient: %s\nN: %s",
                formatC(qval, format = "e", digits = 3),
                formatC(coef_val, format = "e", digits = 2),
                formatC(length(input_df[, "x"]))
              ),
              color = "black",
              size = 2,
              fontface = "italic"
            )
        } else {
          logging::loginfo(
            "Creating boxplot for categorical data, %s vs %s",
            x_label,
            y_label
          )
          input_df["x"] <- lapply(input_df["x"], as.character)
          x_axis_label_names <- unique(input_df[["x"]])
          renamed_levels <- as.character(levels(metadata[
            ,
            x_label
          ]))
          if (length(renamed_levels) == 0) {
            renamed_levels <- x_axis_label_names
          }
          for (name in x_axis_label_names) {
            total <- length(which(input_df[["x"]] == name))
            new_n <- paste(name, " (n=", total, ")", sep = "")
            input_df[which(input_df[["x"]] == name), "x"] <- new_n
            renamed_levels <- replace(
              renamed_levels,
              renamed_levels == name, new_n
            )
          }
          input_df$xnames <-
            factor(input_df[["x"]], levels = renamed_levels)
          temp_plot <- ggplot2::ggplot(
            data = input_df,
            ggplot2::aes(xnames, y)
          ) +
            ggplot2::geom_boxplot(
              ggplot2::aes(fill = x),
              outlier.alpha = 0,
              na.rm = TRUE,
              alpha = 0.5,
              show.legend = FALSE
            ) +
            ggplot2::geom_point(
              ggplot2::aes(fill = x),
              alpha = 0.75,
              size = 1,
              shape = 21,
              stroke = 0.15,
              color = "black",
              position = ggplot2::position_jitterdodge()
            ) +
            paletteer::scale_fill_paletteer_d(palette = "khroma::smoothrainbow")
          temp_plot <- temp_plot + nature_theme(input_df[
            ,
            "x"
          ], y_label) + ggplot2::theme(
            panel.grid.major = ggplot2::element_blank(),
            panel.grid.minor = ggplot2::element_blank(),
            panel.background = ggplot2::element_blank(),
            axis.line = ggplot2::element_line(colour = "black")
          ) +
            ggplot2::xlab(x_label) + ggplot2::ylab(y_label) +
            ggplot2::theme(legend.position = "none") +
            ggplot2::annotate(
              geom = "text",
              x = Inf,
              y = Inf,
              hjust = 1,
              vjust = 1,
              label = sprintf(
                "FDR: %s\nCoefficient: %s\nValue: %s",
                formatC(qval, format = "e", digits = 3),
                formatC(coef_val, format = "e", digits = 2),
                results_value
              ),
              color = "black",
              size = 2,
              fontface = "italic"
            )
        }
        stdout <- capture.output(print(temp_plot), type = "message")
        if (length(stdout) > 0) {
          logging::logdebug(stdout)
        }
        if (save_scatter) {
          saved_plots[[label]][[count]] <- temp_plot
        } else if (count <= max_pngs) {
          saved_plots[[label]][[count]] <- temp_plot
        }
        count <- count + 1
      }
      invisible(dev.off())
      for (plot_number in seq(1, min((count - 1), max_pngs))) {
        png_file <-
          file.path(
            figures_folder,
            paste0(
              substr(
                basename(plot_file),
                1, nchar(basename(plot_file)) - 4
              ), "_", plot_number,
              ".png"
            )
          )
        png(png_file,
          res = 300,
          width = 960,
          height = 960
        )
        stdout <-
          capture.output(print(saved_plots[[label]][[plot_number]]))
        invisible(dev.off())
      }
      if (save_scatter) {
        names(saved_plots[[label]]) <- make.names(output_df_all[
          data_index,
          "feature"
        ], unique = TRUE)
      } else {
        saved_plots[[label]] <- NULL
      }
      metadata_number <- metadata_number + 1
    }
    return(saved_plots)
  }

`%!in%` <- negate(`%in%`)
```

## Training Cohort
### Training Cohort Analysis
#### Univariate Clinical Statistics
```{r, univariate-clinical}
# Build table one for original cohort
tableone_nocovid_df <-
  micu_new_nocovid_oc %>%
  left_join(cri_rxmar_abx_long, by = "unique_id") %>%
  mutate(across(Cephalosporins:Doxycycline, ~ str_to_title(.))) %>% 
  mutate(across(Cephalosporins:Doxycycline, ~ replace_na(., "Unchecked"))) %>%
  mutate(across(Cephalosporins:Doxycycline, ~ as.factor(.))) %>% 
  mutate(across(Cephalosporins:Doxycycline, ~ factor(., levels = c("Unchecked", "Checked")))) %>% 
  mutate(across(Hypertension:Tuberculosis, ~ factor(., levels = c("Unchecked", "Checked")))) %>% 
  mutate(across(Acute.respiratory.distress.syndrome:Newly.diagnosed.solid.malignancy, ~ factor(., levels = c("Unchecked", "Checked")))) %>%
  mutate(across(Myocardial.infract:AIDS, ~ factor(., levels = c("Unchecked", "Checked")))) %>% 
  select(
    age,
    sex.factor,
    bmi,
    race.factor,
    cci_total_sc,
    thirtyday_mortality_overall,
    primary_dx.factor,
    ards.factor,
    sepsis.factor,
    admit_from.factor,
    COVID_upon_admission,
    sofa_score_total,
    ap2_total_score,
    reason_for_intubation.factor,
    reintub_1.factor,
    reintub_2.factor,
    total_ventilator_days,
    icu_los_total,
    hospital_los,
    day_collected,
    Hypertension:`Neuromuscular.disorder`,
    `Peptic.ulcer.disease`,
    `Thyroid.disease`:Tuberculosis,
    `Bacterial.pneumonia`:`Newly.diagnosed.solid.malignancy`,
    `Myocardial.infract`:`AIDS`,
    Penicillins,
    Cephalosporins,
    Carbapenems,
    Vancomycin,
    Metronidazole,
    Macrolides,
    Quinolones,
    other,
    Clindamycin,
    Aminoglycosides,
    Doxycycline,
    `Trimethoprim-Sulfamethoxazole`,
    Rifaximin,
    `diet`,
    dSOFA_admission, 
    dSOFA_stool
  ) %>%
  janitor::clean_names() %>%
  select(-c(
    hypertension:tuberculosis,
    reason_for_intubation_factor:hospital_los
  )) %>%
  replace_na(list(reason_for_intubation_factor = "Not intubated")) %>%
  droplevels()

tableone_nocovid <- CreateTableOne(
  data = tableone_nocovid_df,
  strata = "thirtyday_mortality_overall",
  includeNA = TRUE,
  
)
summary(tableone_nocovid)

# Print tableone
tableone_nocovid_print <-
  print(tableone_nocovid,
    nonnormal = TRUE,
    formatOptions = list(big.mark = ",")
  )

# Export to csv to then load in as a dataframe
write.csv(
  tableone_nocovid_print,
  "./Results/Table_One_30_Days_Mortality_train.csv",
  row.names = TRUE
)

# Clean table for paper
tableone_nocovid_print_clean <-
tableone_nocovid_print %>%
  as.data.frame() %>% 
  rownames_to_column(var = "variable") %>% 
  filter(variable != "thirtyday_mortality_overall...Non.Survivor....") %>% #distinct(variable)
  mutate(
    variable = dplyr::recode(
      variable,
      n = "Number of Patients",
      `age..median..IQR..` = "Age (median [IQR])",
      `sex_factor...Male....` = "Male (%)",
      `bmi..median..IQR..` = "Body Mass Index (median [IQR])",
      `race_factor....` = "Race (%)",
      `X...African.American` = " African American",
      `X...Asian` = "Asian",
      `X...More.than.one.race` = "More than one race",
      `X...White..Hispanic` = "White, Hispanic",
      `X...White..non.Hispanic` = "White, Non-Hispanic",
      `X...NA` = "NA",
      `cci_total_sc..median..IQR..` = "Charlson Comorbidity Index (median [IQR])",
      `primary_dx_factor....` = "Primary admission diagnosis (%)",
      `X...Acute..on.chronic..liver.failure` = "Acute chronic liver failure",
      `X...AMI.dysrhythmia` = "AMI dysrhytmia",
      `X...CHF.cardiogenic.shock` = "CHF cardiogenic shock",
      `X...CNS.pathology` = "CNS pathology",
      `X...GI.hemorrhage` = "GI hemorrhage",
      `X...Metabolic` = "Metabolic",
      `X...Other` = "Other Primary diagnosis",
      `X...Post.operative.observation` = "Post-operative observation",
      `X...Respiratory.failure..AHRF` = "Respiratory failure (AHRF)",
      `X...Respiratory.failure..airway.compromise` = "Respiratory failure, airway compromise",
      `X...Respiratory.failure..ventilatory` = "Respiratory failure, ventilatory",
      `X...Sepsis......septic.shock.` = "Sepsis, septic shock",
      `ards_factor...Yes....` = "Acute respiratory distress syndrome (%)",
      `sepsis_factor...Sepsis....` = "Sepsis (%)",
      `admit_from_factor....` = "Admitted from (%)",
      `X...Cardiology` = "Cardiology",
      `X...ED` = "Emergency Department",
      `X...General.Medicine` = "General Medicine",
      `X...Liver` = "Liver",
      `X...Neurology` = "Nuerology",
      `X...Oncology` = "Oncology",
      `X...OSH` = "Outside Hospital",
      `X...Surgery` = "Surgery",
      `X...NA.1` = "Unknown",
      `covid_upon_admission...No....` = "No Covid upon admission (%)",
      `sofa_score_total..median..IQR..` = "SOFA Score (median [IQR])",
      `ap2_total_score..median..IQR..` = "APACHE II Score (median [IQR])",
      `day_collected..median..IQR..` = "Day From Admission Stool Sample Collected (median [IQR])",
      `bacterial_pneumonia...Checked....` = "Bacterial Pneumonia (%)",
      `fungal_pneumonia...Checked....` = "Fungal Pneumonia (%)",
      `viral_pneumonia...Checked....` = "Viral Pneumonia (%)",
      `chronic_obstructive_pulmonary_disease_copd_1...Checked....` = "Chronic Obstructive Pulmonary Disease (COPD) (%)",
      `asthma_exacerbation...Checked....` = "Asthma exacerbation (%)",
      `lung_lobar_collapse...Unchecked....` = "Lung/lobar collapse (%)",
      `pulmonary_embolism...Checked....` = "Pulmonary embolism (%)",
      `hemoptysis...Checked....` = "Hemoptysis (%)",
      `pancreatitis...Checked....` = "Pancreatitis (%)",
      `infection_genitourinary_system...Checked....` = "Infection, genitourinary system (%)",
      `infection_intra_abdominal...Checked....` = "Infection, Intra-abdominal (%)",
      `infection_soft_tissue...Checked....` = "Infection, soft tissue (%)",
      `infection_cns...Checked....` = "Infection, CNS (%)",
      `hepatic_failure_acute_fullminant...Checked....` = "Hepatic failure, acute fullminant (%)",
      `hepatic_failure_acute_on_chronic...Checked....` = "Hepatic failure, acute on chronic (%)",
      `diabetic_ketoacidosis...Checked....` = "Diabetic ketoacidosis (%)",
      `acute_leukemia...Checked....` = "Acute leukemia (%)",
      `cerebral_vascular_accident_1...Checked....` = "Cerebreal vascular accident (%)",
      `acute_myocardial_infarction_nstemi_stemi...Checked....` = "Acute myocardial infarction (NSTEMI/STEMI) (%)",
      `diffuse_alveolar_hemorrhage...Checked....` = "Diffuse alveolar hemorrhage (%)",
      `decompensated_heart_failure_pulmonary_oedema...Checked....` = "Decompensated heart failure/Pulmonary oedema (%)",
      `pleural_effusion...Checked....` = "Pleural effusion (%)",
      `interstitial_lung_disease_exacerbation...Checked....` = "Interstitial lung disease exacerbation (%)",
      `organizing_pneumonia...Unchecked....` = "Organizing pneumonia (%)",
      `acute_eosinophilic_pneumoniae...Unchecked....` = "Acute eosinophilic pneumoniae (%)",
      `other...Checked....` = "Other (%)",
      `angioedema...Checked....` = "Angioedema (%)",
      `acute_renal_failure...Checked....` = "Acute renal failure (%)",
      `altered_mental_status...Checked....` = "Altered mental status (%)",
      `hypertensive_urgency...Checked....` = "Hypertensive urgency (%)",
      `hypertensive_emergency...Checked....` = "Hypertensive emergency (%)",
      `endocarditis...Checked....` = "Endocarditis (%)",
      `bacteremia...Checked....` = "Bacteremia (%)",
      `gastrointestinal_bleeding...Checked....` = "Gastrointestinal bleeding (%)",
      `hemorrhagic_shock...Checked....` = "Hemorrhagic shock (%)",
      `aspiration...Checked....` = "Aspiration (%)",
      `central_line_associated_blood_steam_infection...Checked....` = "Central line associated blood steam infection (%)",
      `prosthetic_joint_infection...Checked....` = "Prosthetic joint infection (%)",
      `new_onset_atrial_fibrillation...Checked....` = "New onset atrial fibrillation (%)",
      `newly_diagnosed_solid_malignancy...Checked....` = "Newly diagnosed solid malignancy (%)",
      `myocardial_infract...Checked....` = "Myocardial infract (%)",
      `congestive_heart_failure...Checked....` = "Congestive heart failure (%)",
      `peripheral_vascular_disease_cci...Checked....` = "Peripheral vascular disease (%)",
      `cerebrovascular_disease...Checked....` = "Cerebrovascular disease (%)",
      `dementia...Checked....` = "Dementia (%)",
      `chronic_pulmonary_disease...Checked....` = "Chronic pulmonary disease (%)",
      `connective_tissue_disease_1...Checked....` = "Connective tissue disease (%)",
      `ulcer_disease...Checked....` = "Ulcer disease (%)",
      `mild_liver_disease...Checked....` = "Mild liver disease (%)",
      `diabetes_without_complications...Checked....` = "Diabetes (without complications) (%)",
      `diabetes_with_end_organ_damage...Checked....` = "Diabetes (with end organ damage) (%)",
      `hemiplegia...Checked....` = "Hemiplegia (%)",
      `moderate_or_severe_renal_disease...Checked....` = "Moderate or severe renal disease (%)",
      `solid_tumor_non_metastatic...Checked....` = "Solid tumor (non-metastatic) (%)",
      `leukemia...Checked....` = "Leukemia (%)",
      `lymhoma...Checked....` = "Lymphoma (%)",
      `moderate_or_severe_liver_disease...Checked....` = "Moderate or severe liver disease (%)",
      `metastatic_solid_tumor...Checked....` = "Solid tumor (metastatic) (%)",
      `aids...Checked....` = "AIDS (%)",
      `penicillins...Checked....` = "Penicillins (%)",
      `cephalosporins...Checked....` = "Cephalosporins (%)",
      `carbapenems...Checked....` = "Carbapenems (%)",
      `vancomycin...Checked....` = "Vancomycin (%)",
      `metronidazole...Checked....` = "Metronidazole (%)",
      `macrolides...Checked....` = "Macrolides (%)",
      `quinolones...Checked....` = "Quinolones (%)",
      `other_2...Checked....` = "Other Antiobiotics (%)",
      `clindamycin...Checked....` = "Clindamycin (%)",
      `aminoglycosides...Checked....` = "Aminoglycosides (%)",
      `doxycycline...Checked....` = "Doxycycline (%)",
      `trimethoprim_sulfamethoxazole...Checked....` = "Trimethoprim-Sulfamethoxazole (%)",
      `rifaximin...Checked....` = "Rifaximin (%)",
      `diet...npo....` = "Diet (nothing by mouth) (%)",
      `d_sofa_admission..median..IQR..` = "SOFA from admission (median [IQR])",
      `d_sofa_stool..median..IQR..` = "SOFA from Stool Sample (median [IQR])"
    )
  ) %>% 
  column_to_rownames(var = "variable")

# Export to csv to then load in as a dataframe
write.csv(
  tableone_nocovid_print_clean,
  file.path("Results", "Table_One_30_Days_Mortality_train_clean.csv"),
  row.names = TRUE
)

# Import csv as dataframe
tableone_nocovid_csv <-
  read.csv("./Results/Table_One_30_Days_Mortality_train.csv",
    stringsAsFactors = FALSE
  )

# Filter for only p-values <= 0.3 to then include in multi-variable model
tableone_pval_filt <- tableone_nocovid_csv %>%
  dplyr::rename(variable = X) %>%
  mutate(
    p = ifelse(p == "<0.001", 0.001, p),
    p = as.numeric(p)
  ) %>%
  # dplyr::slice(2:5, 11, 26, 27, 38:40, 100:112) %>%
  filter(!grepl(variable, pattern = "^\\s")) %>% 
  janitor::clean_names()

tableone_pval_filt_vars <- tableone_pval_filt %>%
  filter(variable != "n") %>%
  select(variable) %>%
  mutate(
    variable = as.character(variable),
    variable = gsub(
      x = variable,
      pattern = "\\s\\(median \\[IQR\\]\\)|\\s\\(%\\)| = Yes| = [Cc]hecked| = Male| = [Uu]nchecked| = npo| = Sepsis| = None",
      fixed = FALSE,
      replacement = ""
    )
  ) %>%
  filter(variable %!in% c("thirtyday_mortality_overall = Non-Survivor", "covid_upon_admission = No")) %>%
  pull(variable)

tableone_nocovid_df_filt <-
  tableone_nocovid_df[, tableone_pval_filt_vars]

tableone_nocovid_df_filt <- tableone_nocovid_df_filt %>%
  bind_cols(
    micu_new_nocovid_oc %>% ungroup() %>%
      left_join(cri_rxmar_abx_long, by = "unique_id") %>%
      mutate(across(
        Cephalosporins:Quinolones, ~ replace_na(., "unchecked")
      )) %>%
      mutate(across(
        Cephalosporins:Quinolones, ~ as.factor(.)
      )) %>%
      select(unique_id, thirtyday_mortality_overall)
  ) %>%
  relocate(unique_id, .before = NULL) %>%
  mutate_all(as.character) %>% 
  pivot_longer(
    !c(unique_id:day_collected, thirtyday_mortality_overall),
    names_to = "variable",
    values_to = "value"
  ) %>%
  mutate(
    value = as.character(value),
    value = ifelse(value %in% c("Checked", "checked", "diet"), 1, 0)
  ) %>% # diet = 1, npo = 0
  pivot_wider(names_from = "variable", values_from = "value") %>% 
  mutate(age = as.numeric(age),
         bmi = as.numeric(bmi),
         cci_total_sc = as.numeric(cci_total_sc),
         sofa_score_total = as.numeric(sofa_score_total),
         ap2_total_score = as.numeric(ap2_total_score),
         day_collected = as.numeric(day_collected)) %>% 
  mutate_if(is.character, as.factor)
```

#### Univariate -Omics Statistics: Shannon Diversity, Enterococcus/Enterobacterales Relative Abundance
```{r, univariate-omics}
t_metaphlan_micu_nocovid <- metaphlan %>%
  mutate(taxid = as.character(taxid)) %>%
  ungroup() %>%
  right_join(
    micu_new_nocovid_oc %>% # contains both MICU and HD information
      ungroup() %>%
      select(db, ID, shotgunSeq_id, metabolomicsID, sepsis.factor) %>%
      distinct(shotgunSeq_id, .keep_all = TRUE),
    by = "shotgunSeq_id"
  ) %>%
  ungroup() %>%
  select(shotgunSeq_id, metabolomicsID, taxid, db, pctseqs, Total) %>%
  distinct() %>%
  ungroup() %>%
  mutate(pctseqs = as.numeric(pctseqs)) %>%
  filter(pctseqs >= 0.0001) %>%
  group_by(shotgunSeq_id) %>%
  dplyr::add_count(taxid, name = "totalSp") %>%
  mutate(
    seq_id_count = length(unique(shotgunSeq_id)),
    spPres = totalSp / seq_id_count
  ) %>%
  filter(spPres >= 0.10) %>%
  select(-c(Total, seq_id_count, spPres, totalSp)) %>%
  group_by(shotgunSeq_id) %>%
  mutate(pctseqs = pctseqs / sum(pctseqs))

t_metaphlan_micu_nocovid_mat <- t_metaphlan_micu_nocovid %>%
  distinct() %>%
  left_join(metaphlan %>% select(taxid, Species) %>% mutate(taxid = as.character(taxid)),
    relationship = "many-to-many"
  ) %>%
  pivot_wider(
    id_cols = c(shotgunSeq_id),
    names_from = "Species",
    values_from = "pctseqs",
    values_fill = 0,
    values_fn = sum
  ) %>%
  column_to_rownames(var = "shotgunSeq_id")

micu_nocovid_first_samps_omics <- micu_new_nocovid_oc %>%
  left_join(t_metaphlan_micu_nocovid, relationship = "many-to-many") %>%
  left_join(metab_quant_imp_tot_mM, relationship = "many-to-many")

micu_nocovid_first_samps_omics_light <-
  micu_nocovid_first_samps_omics %>%
  select(
    unique_id,
    shotgunSeq_id,
    metabolomicsID,
    taxid,
    pctseqs,
    compound,
    mvalue__mM,
    thirtyday_mortality_overall
  ) %>%
  pivot_wider(
    id_cols = c(
      unique_id,
      shotgunSeq_id,
      metabolomicsID,
      compound,
      mvalue__mM,
      thirtyday_mortality_overall
    ),
    names_from = "taxid",
    values_from = "pctseqs",
    values_fill = 0
  ) %>%
  pivot_longer(
    !c(
      unique_id,
      shotgunSeq_id,
      metabolomicsID,
      compound,
      mvalue__mM,
      thirtyday_mortality_overall
    ),
    names_to = "taxid",
    values_to = "pctseqs"
  )

# Create dataframe for all phylogentic levels of interest
phylo_rel_abd <- t_metaphlan_micu_nocovid %>%
  pivot_wider(
    id_cols = c(shotgunSeq_id, metabolomicsID, db),
    names_from = "taxid",
    values_from = "pctseqs",
    values_fill = 0
  ) %>%
  pivot_longer(!c(shotgunSeq_id, metabolomicsID, db),
    names_to = "taxid",
    values_to = "pctseqs"
  ) %>%
  left_join(
    micu_new_nocovid_oc %>%
      ungroup() %>%
      select(shotgunSeq_id, thirtyday_mortality_overall) %>%
      distinct(shotgunSeq_id,
        .keep_all = TRUE,
        by = "shotgunSeq_id"
      )
  ) %>%
  left_join(taxdmp %>% mutate(taxid = as.character(taxid))) %>%
  mutate(Species = paste(Kingdom, Phylum, Class, Order, Family, Genus, Species, sep = "|")) %>% 
  filter(grepl(pattern = "Enterococcus|Enterobacterales", x = Species)) %>%
  mutate(organism = case_when(
    grepl(pattern = "Enterococcus", x = Species) ~ "Enterococcus",
    grepl(pattern = "Enterobacterales", x = Species) ~ "Enterobacterales"
  )) %>%
  drop_na(organism) %>%
  select(shotgunSeq_id, thirtyday_mortality_overall, organism, pctseqs) %>%
  group_by(shotgunSeq_id, thirtyday_mortality_overall, organism) %>%
  summarise(pctseqs = sum(pctseqs)) %>%
  ungroup() %>%
  pivot_wider(
    names_from = "organism",
    values_from = "pctseqs",
    values_fill = 0
  ) %>%
  pivot_longer(
    !c(shotgunSeq_id, thirtyday_mortality_overall),
    names_to = "organism",
    values_to = "pctseqs"
  ) %>%
  mutate(thirtyday_mortality_overall = factor(
    thirtyday_mortality_overall,
    levels = c("Survivor", "Non-Survivor")
  ))

# Obtain stats for all phylogentic levels of interest
rel_abd_alpha_stats <- phylo_rel_abd %>%
  group_by(organism) %>%
  rstatix::wilcox_test(pctseqs ~ thirtyday_mortality_overall) %>%
  rstatix::adjust_pvalue(method = "BH") %>%
  mutate(p.adj = ifelse(p.adj < 0.001, 0.001, round(p.adj, 3)))

symnum.args <-
  list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, Inf),
    symbols = c("****", "***", "**", "*", "ns")
  )

# Alpha Diversity matrix: Shannon
alpha_shannon <-
  vegan::diversity(t_metaphlan_micu_nocovid_mat, index = "shannon") %>%
  as.data.frame() %>%
  rownames_to_column(var = "shotgunSeq_id") %>%
  dplyr::rename("Shannon" = ".")

# Write out shotgunSeq_id list
alpha_shannon %>%
  select(shotgunSeq_id) %>%
  write.csv(., "./Data/shotgunSeq_id_list.csv", row.names = FALSE)

# Enterococcus
set.seed(456)
gg_ecoc_rel_abd <- phylo_rel_abd %>%
  filter(organism == "Enterococcus") %>%
  group_by(organism) %>%
  ggplot(
    .,
    aes(
      x = thirtyday_mortality_overall,
      y = pctseqs,
      color = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  geom_boxplot(
    outlier.colour = NA,
    alpha = 0.35
  ) +
  geom_jitter(
    width = 0.2,
    size = 2.5,
    alpha = 0.65
  ) +
  stat_compare_means(
    comparisons = list(c("Survivor", "Non-Survivor")),
    tip.length = 0.01,
    # symnum.args = symnum.args,
    method.args = list(
      alternative = "two.sided",
      exact = FALSE
    ),
    label.y = c(1.05)
  ) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title.y = et(size = 14, color = "black"),
    axis.title.x = eb(),
    axis.text = et(size = 12, color = "black"),
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5
      ),
    panel.border = eb(),
    axis.line = el(color = "black")
  ) +
  ylab(~ atop(paste(italic("Enterococcus")), paste("MetaPhlAn4 Relative Abundance"))) +
  ggsci::scale_fill_lancet() +
  ggsci::scale_color_lancet() +
  guides(
    fill = guide_legend("Outcome"),
    color = guide_legend("Outcome",
      override.aes = aes(label = "")
    )
  ) +
  scale_y_continuous(
    breaks = seq(0, 1, 0.1),
    expand = expansion(mult = c(0.01, 0.035)),
    labels = scales::percent_format(accuracy = 1)
  ) +
  coord_cartesian(xlim = c(1.1, 1.9))

gg_ecoc_rel_abd

pdf(
  file = "./Results/Enterococcus_Metaphlan_Outcome_30_Days_Mortality_train.pdf",
  height = 6,
  width = 7
)
gg_ecoc_rel_abd
invisible(invisible(dev.off()))

# Enterobacterales
set.seed(456)
gg_ebac_rel_abd <- phylo_rel_abd %>%
  filter(organism == "Enterobacterales") %>%
  group_by(organism) %>%
  ggplot(
    .,
    aes(
      x = thirtyday_mortality_overall,
      y = pctseqs,
      color = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  geom_boxplot(
    outlier.colour = NA,
    alpha = 0.35
  ) +
  geom_jitter(
    width = 0.2,
    size = 2.5,
    alpha = 0.65
  ) +
  stat_compare_means(
    comparisons = list(c("Survivor", "Non-Survivor")),
    tip.length = 0.01,
    # symnum.args = symnum.args,
    method.args = list(
      alternative = "two.sided",
      exact = FALSE
    ),
    label.y = c(1.05)
  ) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title.y = et(size = 14, color = "black"),
    axis.title.x = eb(),
    axis.text = et(size = 12, color = "black"),
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5
      ),
    panel.border = eb(),
    axis.line = el(color = "black")
  ) +
  ylab(~ atop(paste(italic("Enterobacterales")), paste("MetaPhlAn4 Relative Abundance"))) +
  ggsci::scale_fill_lancet() +
  ggsci::scale_color_lancet() +
  guides(
    fill = guide_legend("Outcome"),
    color = guide_legend("Outcome",
      override.aes = aes(label = "")
    )
  ) +
  scale_y_continuous(
    breaks = seq(0, 1, 0.1),
    expand = expansion(mult = c(0.01, 0.035)),
    labels = scales::percent_format(accuracy = 1)
  ) +
  coord_cartesian(xlim = c(1.1, 1.9))

gg_ebac_rel_abd

pdf(
  file = "./Results/Enterobacterales_Metaphlan_Outcome_30_Days_Mortality_train.pdf",
  height = 6,
  width = 7
)
gg_ebac_rel_abd
invisible(invisible(dev.off()))

dominations <- phylo_rel_abd %>%
  group_by(shotgunSeq_id) %>%
  mutate(
    enterococcus_domination = ifelse(organism == "Enterococcus" &
      pctseqs >= 0.30, 1, 0),
    enterobacterales_domination = ifelse(organism == "Enterobacterales" &
      pctseqs >= 0.05, 1, 0)
  ) %>%
  pivot_longer(!c(shotgunSeq_id:pctseqs),
    names_to = "dominations",
    values_to = "outcome"
  ) %>%
  filter(
    grepl(x = organism, pattern = "Enterococcus") &
      grepl(x = dominations, pattern = "enterococcus_domination") |
      grepl(x = organism, pattern = "Enterobacterales") &
        grepl(x = dominations, pattern = "enterobacterales_domination")
  ) %>%
  pivot_wider(!c(organism, pctseqs),
    names_from = "dominations",
    values_from = "outcome"
  )

micu_nocovid_first_samps_omics_light_filt_wide <-
  micu_nocovid_first_samps_omics_light %>%
  left_join(taxdmp %>% mutate(taxid = as.character(taxid))) %>%
  mutate(Genus = paste0(Phylum, "-", Order, "-", Family, "-", Genus, "-", Species)) %>%
  pivot_wider(
    id_cols = c(unique_id, compound, mvalue__mM),
    names_from = "Genus",
    values_from = "pctseqs",
    values_fn = sum
  ) %>%
  relocate(compound, mvalue__mM, .before = unique_id) %>%
  pivot_wider(
    id_cols = c(unique_id:last_col()),
    names_from = "compound",
    values_from = "mvalue__mM"
  ) %>%
  left_join(
    dominations %>%
      select(
        shotgunSeq_id,
        enterococcus_domination,
        enterobacterales_domination
      ) %>%
      left_join(
        micu_new_nocovid_oc %>%
          ungroup() %>%
          select(unique_id, shotgunSeq_id) %>%
          distinct(shotgunSeq_id, .keep_all = TRUE),
        by = "shotgunSeq_id"
      )
  )

# Enterobacterales + Enterococcus
set.seed(456)
gg_ecoc_ebac_rel_abd <- phylo_rel_abd %>%
  mutate(
    comps = paste(thirtyday_mortality_overall, organism, sep = "\n"),
    comps = factor(
      comps,
      levels = c(
        "Survivor\nEnterococcus",
        "Non-Survivor\nEnterococcus",
        "Survivor\nEnterobacterales",
        "Non-Survivor\nEnterobacterales"
      )
    )
  ) %>%
  ggplot(
    .,
    aes(
      x = comps,
      y = pctseqs,
      color = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  geom_boxplot(
    outlier.colour = NA,
    alpha = 0.35
  ) +
  geom_jitter(
    width = 0.2,
    size = 2.5,
    alpha = 0.65
  ) +
  stat_compare_means(
    comparisons = list(
      c("Survivor\nEnterococcus", "Non-Survivor\nEnterococcus"),
      c(
        "Survivor\nEnterobacterales",
        "Non-Survivor\nEnterobacterales"
      )
    ),
    tip.length = 0.05,
    bracket.size = 0.5,
    symnum.args = symnum.args,
    method.args = list(
      alternative = "two.sided",
      exact = FALSE
    ),
    label.y = c(1.05)
  ) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title.y = et(size = 16, color = "black"),
    axis.title.x = eb(),
    axis.text = et(size = 14, color = "black"),
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5
    ),
    panel.border = eb(),
    axis.line = el(color = "black"),
    legend.position = "none"
  ) +
  ylab(paste("MetaPhlAn4 Relative Abundance\n")) +
  ggsci::scale_fill_lancet() +
  ggsci::scale_color_lancet() +
  guides(
    fill = guide_legend("Outcome"),
    color = guide_legend("Outcome",
      override.aes = aes(label = "")
    )
  ) +
  scale_y_continuous(
    breaks = seq(0, 1, 0.1),
    expand = expansion(mult = c(0.01, 0.05)),
    labels = scales::percent_format(accuracy = 1)
  ) +
  coord_cartesian(xlim = c(1.1, 3.9))

gg_ecoc_ebac_rel_abd

pdf(
  file = "./Results/Enterococcus_Enterobacterales_Metaphlan_Outcome_30_Days_Mortality_train.pdf",
  height = 3.5,
  width = 8
)
gg_ecoc_ebac_rel_abd
invisible(dev.off())
```

#### MaAsLin2 Differential Abundance
```{r, maaslin-differential-abundance-analysis}
# Aggregate to Family level
maaslin_mat <- t_metaphlan_micu_nocovid %>%
  distinct() %>%
  left_join(taxdmp %>% mutate(taxid = as.character(taxid))) %>%
  mutate(Family = ifelse(Family == "", str_extract(Species, pattern = "([^\\s]+)"), Family)) %>%
  group_by(shotgunSeq_id, metabolomicsID, db, Family) %>%
  summarise(pctseqs = sum(pctseqs)) %>%
  pivot_wider(
    id_cols = c(shotgunSeq_id),
    names_from = "Family",
    values_from = "pctseqs",
    values_fill = 0,
    values_fn = sum
  ) %>%
  column_to_rownames(var = "shotgunSeq_id")

# Run Maaslin without covariates
set.seed(123)
maaslin_no_covariates <- Maaslin2(
  input_data = maaslin_mat,
  input_metadata = data.frame(
    t_metaphlan_micu_nocovid_mat %>%
      rownames_to_column(var = "shotgunSeq_id") %>%
      select(shotgunSeq_id) %>%
      left_join(micu_new_nocovid_oc %>%
        select(shotgunSeq_id, unique_id)) %>%
      left_join(
        tableone_nocovid_df_filt %>%
          labelled::remove_labels() %>%
          janitor::clean_names() %>%
          mutate(
            race_factor = as.character(race_factor),
            race_factor = ifelse(
              race_factor %in% c("Asian", "More than one race"),
              "Other",
              race_factor
            )
          )
      ) %>%
      column_to_rownames(var = "shotgunSeq_id") %>%
      select(-c(unique_id)) %>%
      mutate(race_factor = as.factor(race_factor))
  ),
  output = "~/Documents/GitHub/Medical_Intensive_Care_Unit/Results/Maaslin_Base",
  min_abundance = 0.001,
  # At least 0.1% abundance
  min_prevalence = 0.10,
  # Taxa found in at least 10% of samples
  min_variance = -Inf,
  normalization = "NONE",
  transform = "NONE",
  analysis_method = "LM",
  max_significance = 0.05,
  # p.adj <= 0.05 (qval = padjust)
  random_effects = NULL,
  fixed_effects = c("thirtyday_mortality_overall"),
  correction = "BH",
  standardize = TRUE,
  cores = 12,
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  heatmap_first_n = 50,
  reference = c("thirtyday_mortality_overall,Survivor")
)

# Run Maaslin2 with Covariates
set.seed(123)
maaslin_model <- Maaslin2(
  input_data = maaslin_mat,
  input_metadata = data.frame(
    t_metaphlan_micu_nocovid_mat %>%
      rownames_to_column(var = "shotgunSeq_id") %>%
      select(shotgunSeq_id) %>%
      left_join(micu_new_nocovid_oc %>%
        select(shotgunSeq_id, unique_id)) %>%
      left_join(
        tableone_nocovid_df_filt %>%
          labelled::remove_labels() %>%
          janitor::clean_names() %>%
          mutate(
            race_factor = as.character(race_factor),
            race_factor = ifelse(
              race_factor %in% c("Asian", "More than one race"),
              "Other",
              race_factor
            )
          )
      ) %>%
      column_to_rownames(var = "shotgunSeq_id") %>%
      select(-c(unique_id)) %>%
      mutate(race_factor = as.factor(race_factor))
  ),
  output = "~/Documents/GitHub/Medical_Intensive_Care_Unit/Results/Maaslin_Covariate",
  min_abundance = 0.001,
  # At least 0.1% abundance
  min_prevalence = 0.10,
  # Taxa found in at least 10% of samples
  min_variance = -Inf,
  normalization = "NONE",
  transform = "NONE",
  analysis_method = "LM",
  max_significance = 0.05,
  # p.adj <= 0.05 (qval = padjust)
  random_effects = NULL,
  fixed_effects = c(
    "thirtyday_mortality_overall",
    "sex_factor",
    "age",
    "cci_total_sc",
    "ards_factor",
    "sepsis_factor",
    "sofa_score_total",
    "day_collected",
    "race_factor",
    "diet"
  ),
  correction = "BH",
  standardize = TRUE,
  cores = 12,
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  heatmap_first_n = 50,
  reference = c("thirtyday_mortality_overall,Survivor", "race_factor,White")
)

# Manual p-value adjustment for specific comparsisons
maaslin2_all_results <- maaslin_model$results

maaslin2_results <-
  maaslin2_all_results %>% filter(metadata == "thirtyday_mortality_overall") # Discard covariate associations

maaslin2_results$qval <-
  p.adjust(maaslin2_results$pval, method = "BH") # FDR correction using 'BH'
```

#### Quant Metabolites Boxplots
```{r, quant-metabolomics-boxplots-total}
metab_quant_converted <- metab_quant_imp_tot_mM %>%
  right_join(
    micu_new_nocovid_oc %>%
      select(metabolomicsID, thirtyday_mortality_overall, sepsis.factor)
  ) %>%
  select(
    metabolomicsID,
    compound,
    mvalue__mM,
    thirtyday_mortality_overall,
    sepsis.factor
  ) %>%
  drop_na(compound)

metab_boxplot <-
  metab_quant_converted %>%
  ungroup() %>%
  mutate(
    class = case_when(
      compound %in% c(
        "taurocholic acid",
        "glycocholic acid",
        "allocholic acid",
        "alpha-muricholic acid",
        "beta-muricholic acid",
        "omega-muricholic acid",
        "ursocholic acid",
        "glycochenodeoxycholic acid",
        "taurochenodeoxycholic acid"
      ) ~ "Conjugated Primary Bile Acid",
      compound %in% c("cholic acid", "chenodeoxycholic acid") ~ "Primary Bile Acid",
      compound %in% c(
        "3-oxolithocholic acid",
        "alloisolithocholic acid",
        "deoxycholic acid",
        "isodeoxycholic acid",
        "lithocholic acid",
        "ursodeoxycholic acid"
      ) ~ "Secondary Bile Acid",
      compound %in% c(
        "threonine",
        "glycine",
        "tyrosine",
        "tyramine",
        "serine",
        "leucine",
        "isoleucine",
        "valine",
        "phenylalanine",
        "alanine",
        "proline",
        "aspartate",
        "methionine",
        "glutamate",
        "lysine",
        "cysteine",
        "tryptophan"
      ) ~ "Amino Acid",
      compound %in% c(
        "acetate",
        "butyrate",
        "succinate",
        "propionate",
        "5-aminovalerate"
      ) ~ "Fatty Acid",
      compound %in% c(
        "kynurenic acid",
        "anthranilic acid",
        "kynurenine",
        "tryptamine"
      ) ~ "Kynurenine Metabolite",
      compound == "desaminotyrosine" ~ "Phenolic Aromatic",
      compound == "niacin" ~ "B-Vitamin",
      TRUE ~ "Indole"
    ),
    compound = case_when(
      class == "Conjugated Primary Bile Acid" ~ paste(str_to_title(compound), "(1˚Conj. BA)"),
      class == "Primary Bile Acid" ~ paste(str_to_title(compound), "(1˚ BA)"),
      class == "Secondary Bile Acid" ~ paste(str_to_title(compound), "(2˚ BA)"),
      class == "Fatty Acid" & compound == "succinate" ~ paste(str_to_title(compound), "(FA)"),
      class == "Fatty Acid" ~ paste(str_to_title(compound), "(SCFA)"),
      class == "Amino Acid" ~ paste(str_to_title(compound), "(AA)"),
      class == "Phenolic Aromatic" ~ paste(str_to_title(compound), "(Phen. Arom.)"),
      class == "Indole" ~ paste(str_to_title(compound), "(Indole)"),
      class == "Kynurenine Metabolite" ~ paste(str_to_title(compound), "(Kyn. Metab.)"),
      class == "B-Vitamin" ~ paste(str_to_title(compound), "(B-Vitamin)")
    )
  ) %>%
  drop_na() %>%
  mutate(
    compound = factor(
      compound,
      levels = c(
        "Acetate (SCFA)",
        "Butyrate (SCFA)",
        "Propionate (SCFA)",
        "Succinate (FA)",
        "5-Aminovalerate (SCFA)",
        "Chenodeoxycholic Acid (1˚ BA)",
        "Cholic Acid (1˚ BA)",
        "Allocholic Acid (1˚Conj. BA)",
        "Alpha-Muricholic Acid (1˚Conj. BA)",
        "Beta-Muricholic Acid (1˚Conj. BA)",
        "Glycochenodeoxycholic Acid (1˚Conj. BA)",
        "Glycocholic Acid (1˚Conj. BA)",
        "Omega-Muricholic Acid (1˚Conj. BA)",
        "Taurochenodeoxycholic Acid (1˚Conj. BA)",
        "Taurocholic Acid (1˚Conj. BA)",
        "Ursocholic Acid (1˚Conj. BA)",
        "3-Oxolithocholic Acid (2˚ BA)",
        "Alloisolithocholic Acid (2˚ BA)",
        "Deoxycholic Acid (2˚ BA)",
        "Isodeoxycholic Acid (2˚ BA)",
        "Lithocholic Acid (2˚ BA)",
        "Ursodeoxycholic Acid (2˚ BA)",
        "Cysteine (AA)",
        "Glycine (AA)",
        "Phenylalanine (AA)",
        "Proline (AA)",
        "Tryptophan (AA)",
        "Tyramine (AA)",
        "Tyrosine (AA)",
        "5-Hydroxyindoleacetate (Indole)",
        "Melatonin (Indole)",
        "Serotonin (Indole)",
        "Indole-3-Acetamide (Indole)",
        "Indole-3-Acetate (Indole)",
        "Indole-3-Lactate (Indole)",
        "Indole (Indole)",
        "Indole-3-Carboxaldehyde (Indole)",
        "Indole-3-Propionate (Indole)",
        "Indole-3-Acrylate (Indole)",
        "Desaminotyrosine (Phen. Arom.)",
        "Anthranilic Acid (Kyn. Metab.)",
        "Kynurenic Acid (Kyn. Metab.)",
        "Kynurenine (Kyn. Metab.)",
        "Tryptamine (Kyn. Metab.)",
        "Niacin (B-Vitamin)"
      )
    ),
    class = factor(
      class,
      levels = c(
        "Fatty Acid",
        "Primary Bile Acid",
        "Conjugated Primary Bile Acid",
        "Secondary Bile Acid",
        "Amino Acid",
        "Indole",
        "Phenolic Aromatic",
        "Kynurenine Metabolite",
        "B-Vitamin"
      )
    )
  )


metab_boxplot_stats <-
  metab_boxplot %>%
  group_by(class, compound) %>%
  rstatix::wilcox_test(
    mvalue__mM ~ thirtyday_mortality_overall,
    p.adjust.method = "none",
    alternative = "two.sided"
  ) %>%
  rstatix::adjust_pvalue(method = "BH") %>%
  rstatix::add_significance(
    "p.adj",
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 0.1, 1),
    symbols = c("****", "***", "**", "*", "0.1", "ns")
  ) %>%
  mutate(
    p.adj = round(p.adj, 2),
    p.adj = ifelse(p.adj < 0.001, "p.adj < 0.001", paste("p.adj = ", round(p.adj, 2)))
  ) %>%
  add_xy_position() %>%
  mutate(y.position = log(y.position, base = 10) * 1.25)

# Boxplot for all compounds
set.seed(123) # for consistent jittering of points
gg_metab_boxplot <-
  ggboxplot(
    metab_boxplot,
    x = "thirtyday_mortality_overall",
    y = "mvalue__mM",
    fill = "thirtyday_mortality_overall",
    color = "thirtyday_mortality_overall",
    alpha = 0.65,
    outlier.shape = NA,
    facet.by = c("class", "compound")
  ) +
  theme(
    legend.text = et(size = 12, color = "black"),
    legend.title = et(size = 14, color = "black"),
    axis.text.x = eb(),
    axis.title.x = eb(),
    axis.title.y = et(size = 12, color = "black"),
    panel.border = eb(),
    strip.background = er(colour = "white", fill = "white"),
  ) +
  geom_hline(yintercept = 0) +
  geom_segment(aes(
    x = 0.35,
    y = 0,
    xend = 0.35,
    yend = Inf
  )) +
  facet_wrap(~compound, scales = "fixed") +
  stat_pvalue_manual(metab_boxplot_stats,
    label = "p.adj",
    tip.length = 0.015
  ) +
  geom_point(
    data = metab_boxplot,
    aes(x = thirtyday_mortality_overall, y = mvalue__mM, color = thirtyday_mortality_overall),
    position = position_jitter(width = 0.2),
    size = 2,
    alpha = 0.65
  ) +
  ggsci::scale_fill_lancet() +
  ggsci::scale_color_lancet() +
  scale_y_log10(
    limits = c(0.0001, 1000),
    labels = c("0.0001", "0.001", "0.01", "0.1", "1", "10", "100", "1000"),
    breaks = c(0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000),
    expand = expansion(mult = c(0.1, 0.2))
  ) +
  ylab("Concentration (mM)\n") +
  labs(
    color = "Outcome",
    fill = "Outcome"
  )

gg_metab_boxplot

cairo_pdf(
  filename = "./Results/Quant_Boxplots_Total_30_Days_Mortality_train.pdf",
  width = 14,
  height = 10,
  onefile = FALSE
)
gg_metab_boxplot
invisible(invisible(dev.off()))
```

#### Cutpoint Analysis: Metabolic Dysbiosis Score
```{r, cutpoint-analysis-mds}
# Cutpoint dataframe
cutpoints_df <- metab_quant_imp_tot_mM %>%
  pivot_wider(
    id_cols = c(metabolomicsID),
    names_from = "compound",
    values_from = "mvalue__mM"
  ) %>%
  group_by(metabolomicsID) %>%
  pivot_longer(!c(metabolomicsID),
    names_to = "compound",
    values_to = "mvalue__mM"
  ) %>%
  right_join(micu_new_nocovid_oc %>% select(metabolomicsID, thirtyday_mortality_overall)) %>%
  group_by(compound) %>%
  mutate(n = length(compound)) %>%
  ungroup() %>%
  mutate(p = length(unique(metabolomicsID))) %>%
  mutate(
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  )

# Optimal cutpoints
# Create function to handle any errors during map function
safe_cutpointr <- possibly(.f = cutpointr, otherwise = "Error")

# set.seed(123)
cutpoints <-
  cutpoints_df %>%
  group_by(compound) %>%
  group_map(
    ~ safe_cutpointr(
      .,
      mvalue__mM,
      thirtyday_mortality_overall_class,
      compound,
      method = maximize_metric,
      metric = youden,
      pos_class = 1,
      neg_class = 0,
      boot_runs = 10,
      boot_stratify = TRUE,
      use_midpoints = TRUE,
      na.rm = TRUE
    ),
    .keep = TRUE
  )

cutpoints_unnest <- cutpoints %>%
  map_df(as_tibble)

cutpoints_unnest %>%
  select(
    compound = subgroup,
    direction,
    optimal_cutpoint_mM = optimal_cutpoint
  ) %>%
  write.csv(., "./Results/MDScore_Cutpoints_train.csv")

# Summary table
cutpoints_unnest_summary <-
  cutpoints_unnest %>%
  group_by(subgroup, pos_class) %>%
  summarize(top_auc = max(AUC)) %>%
  filter(top_auc == max(top_auc)) %>%
  arrange(-top_auc)

# Plot top results
cutpoints_unnest %>%
  mutate(
    tvalue = as.numeric(
      str_extract(string = pos_class, pattern = "[0-9]\\.[0-9]+|[0-9]+")
    ),
    variable = gsub("\\s<=.*", "", pos_class)
  ) %>%
  separate(subgroup,
    c("group1", "group2"),
    sep = "__",
    remove = FALSE
  ) %>%
  select(-boot) %>%
  mutate(group_ratio = if_else(!is.na(group2), paste(group1, group2, sep = " : "), group1)) %>%
  arrange(desc(AUC)) %>%
  group_by(pos_class) %>%
  group_by(tvalue, variable, subgroup, group_ratio, pos_class) %>%
  summarize(top_auc = max(AUC)) %>%
  ungroup() %>%
  arrange(desc(top_auc)) %>%
  group_by(tvalue, pos_class) %>%
  arrange(pos_class, tvalue, subgroup, group_ratio) %>%
  droplevels() %>%
  mutate(variable = "Predicting Outcomes: Survivor vs Non-Survivor") %>%
  ggplot() +
  geom_bar(
    aes(
      x = reorder(group_ratio, -top_auc),
      y = top_auc,
      fill = group_ratio
    ),
    stat = "identity",
    position = "dodge"
  ) +
  geom_hline(yintercept = 0.9) +
  geom_hline(yintercept = 0.8) +
  geom_hline(yintercept = 0.7) +
  shadowtext::geom_shadowtext(
    aes(
      x = group_ratio,
      y = top_auc / 2.5,
      angle = 90,
      label = group_ratio
    ),
    size = 4
  ) +
  shadowtext::geom_shadowtext(aes(
    x = group_ratio,
    y = top_auc * 1.015,
    label = round(top_auc, 2)
  )) +
  theme_bw() +
  theme(
    panel.grid.minor = eb(),
    panel.grid.major.x = eb(),
    strip.text = et(size = 14, color = "black"),
    axis.text.y = et(size = 12, color = "black"),
    axis.text.x = eb(),
    axis.ticks.x = eb(),
    axis.title.y = et(size = 14, color = "black"),
    axis.title.x = eb(),
    legend.title = et(size = 14, color = "black"),
    legend.text = et(
      size = 12,
      color = "black",
      hjust = 0
    ),
    legend.position = "none"
  ) +
  ggsci::scale_fill_igv() +
  xlab("\nMetabolite Concentration") +
  ylab("AUC \n") +
  scale_y_continuous(limits = c(0, 1), breaks = seq(0, 1, 0.1)) +
  guides(fill = guide_legend(ncol = 1)) +
  labs(fill = "Metabolite Concentration") +
  ggtitle("Predicting MICU Outcomes: Survivor vs Non-Survivor") +
  facet_grid(~variable)

ggsave(
  "./Results/Cutpoint_AUC_30_Days_Mortality_train.pdf",
  height = 6,
  width = 12,
  units = "in"
)

# Build dataframe to use cutpoints
cutpoints_results <-
  cutpoints_df %>%
  left_join(
    cutpoints_unnest %>%
      dplyr::rename(compound = subgroup) %>%
      select(compound, direction, optimal_cutpoint)
  ) %>%
  mutate(
    cutpoint_prediction = case_when(
      direction == "<=" & mvalue__mM <= optimal_cutpoint ~ 1,
      direction == "<=" &
        mvalue__mM > optimal_cutpoint ~ 0,
      direction == ">=" &
        mvalue__mM >= optimal_cutpoint ~ 1,
      direction == ">=" &
        mvalue__mM < optimal_cutpoint ~ 0
    )
  ) %>%
  group_by(metabolomicsID, compound) %>%
  mutate(md_score = sum(cutpoint_prediction)) %>%
  dplyr::slice(1) %>%
  pivot_wider(
    id_cols = c(thirtyday_mortality_overall, metabolomicsID),
    names_from = "compound",
    values_from = "cutpoint_prediction"
  ) %>%
  column_to_rownames(var = "metabolomicsID") %>%
  relocate(thirtyday_mortality_overall, .after = last_col()) %>%
  mutate(thirtyday_mortality_overall = as.factor(thirtyday_mortality_overall))

# Use Ridge to find most predictive cutpoints
set.seed(564)
cutpoint_ridge <-
  cv.glmnet(
    x = cutpoints_results %>% select(-thirtyday_mortality_overall) %>% as.matrix(),
    y = factor(cutpoints_results$thirtyday_mortality_overall, labels = c(0, 1)),
    family = "binomial",
    type.measure = "auc",
    nfolds = 10,
    alpha = 0,
  )

# Find optimal lambda value that minimizes test MSE
cutpoint_best_lambda <- cutpoint_ridge$lambda.min
cutpoint_best_lambda

# Produce plot of test MSE by lambda value
plot(cutpoint_ridge)

# Find coefficients of best model
cutpoint_best_ridge <-
  glmnet(
    x = cutpoints_results %>% select(-thirtyday_mortality_overall) %>% as.matrix(),
    y = factor(cutpoints_results$thirtyday_mortality_overall, labels = c(0, 1)),
    family = "binomial",
    standardize = FALSE,
    alpha = 0,
    lambda = cutpoint_best_lambda
  )

# Save MDScore CSV in order of ridge regression importance
as.matrix(coef(cutpoint_best_ridge)) %>%
  as.data.frame() %>%
  rownames_to_column(var = "compound") %>%
  filter(compound != "(Intercept)") %>%
  arrange(desc(abs(s0))) %>%
  dplyr::rename(beta_coefficient = s0) %>% 
  left_join(
    cutpoints_unnest %>%
      select(
        compound = subgroup,
        direction,
        optimal_cutpoint_mM = optimal_cutpoint
      )
  ) %>% 
write.csv(., "./Results/MDScore_Cutpoints_train_ridge_order.csv")

min_loop_cmpds <- 2
max_loop_cmpds <-
  ncol(cutpoints_results %>% select(-thirtyday_mortality_overall) %>% as.matrix()) - 1

roc_loop_df <- NULL

for (i in seq(min_loop_cmpds, max_loop_cmpds, 1)) {
  # i = 15
  top_roc_cmpds <- as.matrix(coef(cutpoint_best_ridge)) %>%
    as.data.frame() %>%
    rownames_to_column(var = "compound") %>%
    filter(compound != "(Intercept)") %>%
    arrange(desc(abs(s0))) %>%
    slice_max(abs(s0), n = i)

  top_roc_cmpds_temp <-
    paste0(
      top_roc_cmpds %>% mutate(compound = gsub(
        x = compound,
        pattern = " ",
        replacement = "-"
      )) %>% pull(compound),
      sep = "_",
      collapse = ""
    )

  top_roc_cmpds_temp2 <-
    paste0(
      top_roc_cmpds %>% mutate(compound = gsub(
        x = compound,
        pattern = " ",
        replacement = "-"
      )) %>% pull(compound),
      sep = "\n",
      collapse = ""
    )

  cutpoints_results_var_slct <-
    cutpoints_df %>%
    filter(compound %in% top_roc_cmpds$compound) %>%
    left_join(
      cutpoints_unnest %>%
        dplyr::rename(compound = subgroup) %>%
        select(compound, direction, optimal_cutpoint)
    ) %>%
    mutate(
      cutpoint_prediction = case_when(
        direction == "<=" & mvalue__mM <= optimal_cutpoint ~ 1,
        direction == "<=" & mvalue__mM > optimal_cutpoint ~ 0,
        direction == ">=" & mvalue__mM >= optimal_cutpoint ~ 1,
        direction == ">=" & mvalue__mM < optimal_cutpoint ~ 0
      )
    ) %>%
    group_by(metabolomicsID, thirtyday_mortality_overall) %>%
    summarize(md_score = sum(cutpoint_prediction))

  # ROC curve for MD Score using training data
  pROC_obj <- pROC::roc(
    cutpoints_results_var_slct$thirtyday_mortality_overall,
    cutpoints_results_var_slct$md_score,
    smoothed = FALSE,
    ci = TRUE,
    plot = FALSE,
    auc.polygon = TRUE,
    best.method = TRUE,
    print.auc = TRUE,
    print.auc.col = "black",
    col = "#2F472F",
    auc.polygon.border = "black",
    auc.polygon.col = "gray65",
    print.thres.best.method = "youden"
  )

  loop_auc <- pROC::auc(
    pROC::roc(
      cutpoints_results_var_slct$thirtyday_mortality_overall,
      cutpoints_results_var_slct$md_score,
      smoothed = FALSE,
      ci = TRUE,
      plot = FALSE,
      auc.polygon = TRUE,
      best.method = TRUE,
      print.auc = TRUE,
      print.auc.col = "black",
      col = "#2F472F",
      auc.polygon.border = "black",
      auc.polygon.col = "gray65",
      print.thres.best.method = "youden"
    )
  )[1]

  coordinates <-
    cbind(data.frame(auc = loop_auc), coords(
      pROC_obj,
      "best",
      ret = c("auc", "threshold", "accuracy", "sens", "spec", "ppv", "npv")
    ))

  pROC::roc(
    cutpoints_results_var_slct$thirtyday_mortality_overall,
    cutpoints_results_var_slct$md_score,
    smooth = FALSE,
    ci = TRUE,
    plot = TRUE,
    auc.polygon = TRUE,
    print.auc = TRUE,
    print.auc.col = "black",
    col = "#2F472F",
    auc.polygon.border = "black",
    auc.polygon.col = "gray65",
    print.thres.best.method = "youden"
  )

  text(
    paste("ACC:", round(coordinates$accuracy, 3) * 100, "%"),
    x = 0.5,
    y = 0.45,
    adj = 0
  )
  text(paste("PPV:", round(coordinates$ppv, 2)),
    x = 0.5,
    y = 0.41,
    adj = 0
  )
  text(paste("NPV:", round(coordinates$npv, 2)),
    x = 0.5,
    y = 0.37,
    adj = 0
  )
  text(
    paste("Threshold:", round(coordinates$threshold, 2)),
    x = 0.5,
    y = 0.33,
    adj = 0
  )
  text(
    paste("i:", i, "\n", top_roc_cmpds_temp2),
    x = 1.1,
    y = 0.6,
    adj = 0
  )

  roc_plot <- grDevices::recordPlot()

  cairo_pdf(
    paste0(
      "./Results/ROC_curve_cutpoint_30_Day_Mortality_",
      i,
      ".pdf"
    ),
    width = 8,
    height = 6
  )
  pROC_obj <- pROC::roc(
    cutpoints_results_var_slct$thirtyday_mortality_overall,
    cutpoints_results_var_slct$md_score,
    smooth = FALSE,
    ci = TRUE,
    plot = TRUE,
    auc.polygon = TRUE,
    print.auc = TRUE,
    print.auc.col = "black",
    col = "#2F472F",
    auc.polygon.border = "black",
    auc.polygon.col = "gray65",
    print.thres.best.method = "youden"
  )
  text(
    paste("ACC:", round(coordinates$accuracy, 3) * 100, "%"),
    x = 0.5,
    y = 0.45,
    adj = 0
  )
  text(paste("PPV:", round(coordinates$ppv, 2)),
    x = 0.5,
    y = 0.41,
    adj = 0
  )
  text(paste("NPV:", round(coordinates$npv, 2)),
    x = 0.5,
    y = 0.37,
    adj = 0
  )
  text(
    paste("Threshold:", round(coordinates$threshold, 2)),
    x = 0.5,
    y = 0.33,
    adj = 0
  )
  text(
    paste("i:", i, "\n", top_roc_cmpds_temp2),
    x = 1.1,
    y = 0.6,
    adj = 0
  )

  invisible(dev.off())

  # Build looped csv of all model metrics and their compounds
  roc_loop_df <- base::rbind(
    roc_loop_df,
    coordinates %>%
      mutate(
        i = i,
        compounds = top_roc_cmpds_temp
      ) %>%
      mutate(across(auc:npv, \(x) round(x, 3))) %>%
      rowid_to_column() %>%
      separate(
        col = compounds,
        into = paste("compound", seq(1, max_loop_cmpds, 1)),
        sep = "_"
      )
  )
}

# Output to CSV files
write.csv(
  roc_loop_df,
  "./Results/ROC_Results_Ridge_Cutpoint_train.csv"
)

cutpoints_roc_loop <-
  roc_loop_df %>%
  group_by(i) %>% # i is coming from the last iteration of the loop above
  slice_max(threshold) %>% # This is because the i = 2 returns both Inf and -Inf values for the threshold
  pivot_longer(
    cols = !c(i, rowid, `compound 1`:last_col()),
    names_to = "model_parameter",
    values_to = "model_value"
  ) %>%
  pivot_longer(
    cols = !c(i, rowid, model_parameter, model_value),
    names_to = "compound_id",
    values_to = "compound"
  ) %>%
  drop_na() %>%
  filter(grepl(compound, pattern = "\\w+")) %>%
  mutate(compound = gsub(
    x = compound,
    pattern = "-acid",
    replacement = " acid"
  )) %>%
  dplyr::rename("number_of_compounds" = i) %>%
  filter(model_parameter == "threshold") %>% 
  group_by(number_of_compounds, compound) %>% 
  dplyr::slice(1) %>% 
  ungroup() %>% 
  left_join(
    cutpoints_df %>% select(
      metabolomicsID,
      compound,
      mvalue__mM,
      thirtyday_mortality_overall,
      thirtyday_mortality_overall_class
    )
  ) %>% 
  left_join(
    cutpoints_unnest %>%
      dplyr::rename(compound = subgroup) %>%
      select(compound, direction, optimal_cutpoint)
  ) %>%
  mutate(
    cutpoint_prediction = case_when(
      direction == "<=" & mvalue__mM <= optimal_cutpoint ~ 1,
      direction == "<=" & mvalue__mM > optimal_cutpoint ~ 0,
      direction == ">=" & mvalue__mM >= optimal_cutpoint ~ 1,
      direction == ">=" & mvalue__mM < optimal_cutpoint ~ 0
    )
  ) %>%
  group_by(number_of_compounds, metabolomicsID, model_parameter, model_value, thirtyday_mortality_overall) %>%
  summarize(md_score = sum(cutpoint_prediction)) %>% 
  mutate(grouped_md_score = ifelse(md_score > model_value, "High Score", "Low Score"))

# Kaplan Meier
km_nocovid <- micu_new_nocovid_oc %>%
  select(
    unique_id,
    sampleid,
    metabolomicsID,
    days_until_death_overall,
    censoring_thirtyday_mortality_overall,
    thirtyday_mortality_overall
  ) %>%
  ungroup() %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>% # Non-Survivor is 1
  left_join(
    cutpoints_roc_loop %>% select(number_of_compounds, metabolomicsID, grouped_md_score, md_score)
  ) %>%
  mutate(grouped_md_score_class = factor(
    grouped_md_score,
    levels = c("Low Score", "High Score"),
    labels = c(1, 0)
  ))                      # Low Score = 1, High Score = 0, due to area = Low Score - High Score)

km_loop <-
  km_nocovid %>%
  left_join(
    tableone_nocovid_df_filt %>% select(unique_id, age:day_collected, penicillins:last_col())
  ) %>%
  group_by(number_of_compounds) %>%
  group_map(~ (
    rmst2(
      time = .x$surv_days,
      status = .x$thirtyday_mortality_overall_class,
      arm = .x$grouped_md_score_class,
      # covariates = c(
      #   .x$age,
      #   .x$sex_factor,
      #   # .x$race_factor, # Check if commas are the problem
      #   .x$cci_total_sc,
      #   .x$ards_factor,
      #   .x$sepsis_factor,
      #   .x$sofa_score_total,
      #   .x$ap2_total_score,
      #   .x$day_collected,
      #   .x$diet
      # ),
      tau = 30
    )
  )) 

km_loop_df <- as.data.frame(do.call(rbind, km_loop))

km_loop_df2 <- NULL

for (i in seq(1,29,1)){
  km_loop_df2 <- janitor::clean_names(rbind(km_loop_df2, km_loop_df[["unadjusted.result"]][i][1][[1]], i))
}

km_loop_df2 <-
  as.data.frame(km_loop_df2) %>% 
  rownames_to_column(var = "variable") %>% 
  filter(variable %in% c("rmst_arm_1_arm_0",
                         "rmst_arm_1_arm_0_2_2",
                         "rmst_arm_1_arm_0_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2",
                         "rmst_arm_1_arm_0_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2_2")) %>% 
  rownames_to_column(var = "number_of_compounds") %>% 
  mutate(number_of_compounds = as.numeric(number_of_compounds) + 1)

# Plot results
# Choose optimal model (based on Accuracy, AUC, NPV, PPV, and complexity)
optimal_components <- 13

hline_dat <-
  data.frame(
    model_parameter = c("AUC", "Accuracy", "Sensitivity", "Specificity"),
    value = c(
      # -1 because of i = 1 being 2 compounds etc.
      roc_loop_df$auc[i = optimal_components - 1],
      roc_loop_df$accuracy[i = optimal_components - 1],
      roc_loop_df$sensitivity[i = optimal_components - 1],
      roc_loop_df$specificity[i = optimal_components - 1]
    )
  )

# Plot results
gg_roc_loop_metrics <-
  roc_loop_df %>%
  group_by(i) %>% # i is coming from the last iteration of the loop above
  slice_max(accuracy) %>% # This is because the i = 2 returns both Inf and -Inf values for the threshold
  pivot_longer(
    cols = !c(i, rowid, `compound 1`:last_col()),
    names_to = "model_parameter",
    values_to = "model_value"
  ) %>%
  pivot_longer(
    cols = !c(i, rowid, model_parameter, model_value),
    names_to = "compound_id",
    values_to = "compound"
  ) %>%
  drop_na() %>%
  filter(grepl(compound, pattern = "\\w+")) %>%
  dplyr::rename("number_of_compounds" = i) %>%
  filter(is.finite(model_value)) %>%
  distinct(rowid, number_of_compounds, model_parameter, model_value) %>%
  mutate(model_parameter = ifelse(
    nchar(model_parameter) > 3,
    str_to_title(model_parameter),
    str_to_upper(model_parameter)
  )) %>%
  group_by(model_parameter) %>%
  mutate(
    model_value_lab = ifelse(model_parameter == "Threshold", NA, model_value),
    model_value_lab = sprintf("%.2f", model_value_lab),
    model_value_lab = ifelse(model_value_lab == "NA", NA, model_value_lab)
  ) %>%
  group_by(model_parameter) %>%
  mutate(
    max_parameter = max(model_value),
    max_parameter_color = ifelse(max_parameter == model_value, "max", NA)
  ) %>%
  filter(model_parameter %in% c("AUC", "Accuracy", "Sensitivity", "Specificity")) %>%
  ggpubr::ggdotchart(
    x = "number_of_compounds",
    y = "model_value",
    color = "model_parameter",
    fill = "max_parameter_color",
    label = "model_value_lab",
    add = "segment",
    sorting = "none",
    dot.size = 3.5,
    font.label = list(
      color = "black",
      size = 8,
      hjust = 2.25,
      vjust = -1
    ),
  ) +
  geom_hline(
    data = hline_dat,
    aes(yintercept = value, color = model_parameter),
    alpha = 0.5
  ) +
  annotate(
    "rect",
    xmin = optimal_components-1.5,
    xmax = optimal_components-0.5,
    ymin = -Inf,
    ymax = Inf,
    alpha = 0.2,
    fill = "grey70"
  ) +
  theme(
    legend.position = "right",
    axis.text.x = eb(),
    axis.title.x = eb()
  ) +
  facet_wrap(~model_parameter, ncol = 1) +
  scale_color_manual(values = c(paletteer::paletteer_d("nbapalettes::bulls_city"), "#005076FF")) +
  guides(color = guide_legend(
    title = "Model Parameter",
    keyheight = 1.5,
    keywidth = 1.5
  )) +
  ylab("Model Metric Value\n") +
  scale_y_continuous(
    expand = expansion(mult = c(0, 0.4)),
    breaks = c(0, 0.5, 1)
  )

gg_km_loop <-
  km_loop_df2 %>%
  arrange(number_of_compounds) %>%
  select(-variable) %>% 
  mutate_all(as.numeric) %>%
  ggpubr::ggdotchart(
    .,
    x = "number_of_compounds",
    y = "est",
    sorting = "none",
    size = 3.5,
    color = "p"
  ) +
  geom_errorbar(
    aes(ymax = `upper_95`, ymin = `lower_95`),
    position = position_dodge(width = 0.8),
    width = 0.25
  ) +
  geom_hline(
    data = data.frame(value = km_loop_df2 %>% filter(number_of_compounds == optimal_components) %>% pull(est)),
    aes(yintercept = value), color = "#C75DAAFF",
    alpha = 0.5
  ) +
  annotate(
    "rect",
    xmin = optimal_components - 1.5,
    xmax = optimal_components - 0.5,
    ymin = -Inf,
    ymax = Inf,
    alpha = 0.2,
    fill = "grey70"
  ) +
  theme(
    axis.title.x = eb(),
    axis.text.x = eb(),
    legend.position = "right"
  ) +
  ylab("∆ RMST\n") +
  xlab("") +
  paletteer::scale_color_paletteer_c(trans = "log10", "grDevices::Tropic") +
  labs(color = "p-value") +
  scale_x_discrete(expand = expansion(add = c(0.6, 1)))

gg_roc_loop_compounds <-
  roc_loop_df %>%
  pivot_longer(
    cols = !c(i, rowid, `compound 1`:last_col()),
    names_to = "model_parameter",
    values_to = "model_value"
  ) %>%
  pivot_longer(
    cols = !c(i, rowid, model_parameter, model_value),
    names_to = "compound_id",
    values_to = "compound"
  ) %>%
  drop_na() %>%
  filter(grepl(compound, pattern = "\\w+")) %>%
  dplyr::rename("number_of_compounds" = i) %>%
  distinct(number_of_compounds, compound) %>%
  mutate(
    compound = str_to_title(compound),
    compound = gsub(compound, pattern = "-Acid", replacement = " Acid"),
    compound = factor(compound, levels = rev(c(
      gsub(
        gsub(
          str_to_title(top_roc_cmpds$compound),
          pattern = " ",
          replacement = "-"
        ),
        pattern = "-Acid",
        replacement = " Acid"
      )
    ))),
    number_of_unique_compounds = 1
  ) %>% # This is important to run the loop before this, so it ends with i = 30 for top_roc_cmpds
  ggpubr::ggbarplot(x = "number_of_compounds", y = "number_of_unique_compounds", fill = "compound") +
  theme( # legend.position = c(1.15,0.9981),
    # legend.justification = c("right", "top"),
    # legend.spacing.y = unit(0, "lines"),
    # legend.background = eb(),
    # panel.grid.major.x = el(color = "black")
    legend.position = "right"
  ) +
  annotate(
    "rect",
    xmin = optimal_components-1.5,
    xmax = optimal_components-0.5,
    ymin = -Inf,
    ymax = Inf,
    alpha = 0.2,
    fill = "grey70"
  ) +
  paletteer::scale_fill_paletteer_d(palette = "Polychrome::glasbey") +
  guides(fill = guide_legend(
    title = "",
    keyheight = 0.75,
    keywidth = 1,
    title.vjust = 1,
    ncol = 1
  )) +
  ylab("Number of Compounds Included in ROC\n") +
  xlab("\nNumber of Compounds Included in ROC") +
  scale_y_discrete(expand = expansion(add = c(0.1, 0.1))) +
  scale_x_discrete(expand = expansion(add = c(0.6, 1)))

# Combine plots
cairo_pdf("./Results/ROC_Loop_train.pdf", height = 12, width = 14)

gg_roc_loop_metrics / gg_km_loop/ gg_roc_loop_compounds +
  patchwork::plot_layout(heights = c(1, 0.35, 1.1))

invisible(dev.off())
```

#### Optimal Metabolic Dysbiosis Score
```{r, optimal-mds}
# KM Curves: MD Score
km_nocovid_final <- km_nocovid %>% 
  filter(number_of_compounds == optimal_components)

set.seed(123)
surv_object <-
  Surv(
    time = km_nocovid_final$surv_days,
    event = km_nocovid_final$thirtyday_mortality_overall_class
  )

fit1 <- survfit(surv_object ~ grouped_md_score, data = km_nocovid_final)

ggs <- ggsurvplot(
  fit1,
  data = km_nocovid_final,
  size = 1,
  palette = c("#C45258", "#2F4858"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  legend.labs = c("High MD Score", "Low MD Score")
)

# Change table axis labels
ggs$table <-
  ggs$table + labs(x = NULL, y = NULL) + theme(plot.title = eb()) # risk table

ggs

pdf(
  "./Results/kaplan_meier_roc_loop_30_Day_Mortality_train.pdf",
  height = 6,
  width = 8,
  onefile = FALSE
)
ggs
invisible(dev.off())

# RMST
rmst2(
      time = km_nocovid_final$surv_days,
      status = km_nocovid_final$thirtyday_mortality_overall_class,
      arm = km_nocovid_final$grouped_md_score_class,
      tau = 30
    )

# Boxplot of MD Score
mds_chis <-
  stats::chisq.test(
    km_nocovid_final$thirtyday_mortality_overall,
    km_nocovid_final$md_score
  )

md_violin <-
  ggviolin(
    km_nocovid_final,
    x = "thirtyday_mortality_overall",
    y = "md_score",
    fill = "thirtyday_mortality_overall",
    palette = "lancet",
    add = c("dotplot"),
    add.params = list(binwidth = 0.05)
  ) +
  annotate(
    "text",
    x = 1.5,
    y = 16,
    label = paste0(
      "Chisq",
      "(",
      round(mds_chis$statistic, 3),
      "),",
      " p =",
      scales::scientific(mds_chis$p.value)
    )
  ) +
  annotate(
    "segment", # horizontal line between the two violins
    x = 1,
    xend = 2,
    y = 15.5,
    yend = 15.5
  ) +
  annotate(
    "segment", # vertical segment connecting to horizontal line above survivor group
    x = 1,
    xend = 1,
    y = 15.5,
    yend = 15.2
  ) +
  annotate(
    "segment",
    x = 2,
    xend = 2,
    y = 15.5,
    yend = 15.2
  ) +
  ylab("Metabolic Dysbiosis Score\n") +
  xlab("") +
  guides(fill = guide_legend("30 Day Mortality"))

md_violin

ggsave(
  plot = md_violin,
  filename = "./Results/MDS_Violin_train.pdf",
  height = 6,
  width = 8
)

km_nocovid_final |> 
  group_by(thirtyday_mortality_overall) |> 
  summarise(mean = mean(md_score),
            sd = sd(md_score))

stats::wilcox.test(
    km_nocovid_final$md_score ~ km_nocovid_final$thirtyday_mortality_overall
  )

# gg_mds_chi <- gginference::ggchisqtest(mds_chis, colaccept = "green3", colreject = "red3") # It is highly unlikely that our test statistic would be observed if there were no association between survival outcome and the md score
# gg_mds_chi

# Confusion matrix for 14 compounds: training model metrics
km_nocovid_final2 <- km_nocovid_final %>% 
  mutate(prediction = ifelse(grouped_md_score == "Low Score", "Survivor", "Non-Survivor"))

pROC::roc(km_nocovid_final2$thirtyday_mortality_overall_class,
          km_nocovid_final2$md_score)
# AUC: 0.8563

caret::confusionMatrix(table(
  factor(
    km_nocovid_final2$prediction,
    levels = c("Survivor", "Non-Survivor")
  ),
  factor(
    km_nocovid_final2$thirtyday_mortality_overall,
    levels = c("Survivor", "Non-Survivor")
  )
))

# Confusion Matrix and Statistics
# 
#               
#                Survivor Non-Survivor
#   Survivor           91           13
#   Non-Survivor       11           32
#                                           
#                Accuracy : 0.8367          
#                  95% CI : (0.7669, 0.8925)
#             Sensitivity : 0.8922          
#             Specificity : 0.7111 
                 
```

#### Cutpoint Analysis: Shannon Diversity
```{r, cutpoint-analysis-shannon}
# Cutpoint dataframe
cutpoints_df_shannon <- alpha_shannon %>%
  left_join(
    micu_new_nocovid_oc %>%
      ungroup() %>%
      select(unique_id, shotgunSeq_id, thirtyday_mortality_overall) %>%
      distinct(shotgunSeq_id, .keep_all = TRUE),
    by = "shotgunSeq_id"
  ) %>%
  mutate(
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  )

# Optimal cutpoints
set.seed(123456)
cutpoints_shannon <-
  cutpoints_df_shannon %>%
  group_map(
    ~ safe_cutpointr(
      .,
      Shannon,
      thirtyday_mortality_overall_class,
      method = maximize_metric,
      metric = youden,
      pos_class = 0,
      neg_class = 1,
      boot_runs = 100,
      use_midpoints = TRUE,
      na.rm = T
    ),
    .keep = TRUE
  )

cutpoints_unnest_shannon <- cutpoints_shannon %>%
  map_df(as_tibble)

# Summary table
cutpoints_unnest_summary_shannon <-
  cutpoints_unnest_shannon %>%
  group_by(pos_class, optimal_cutpoint) %>%
  summarize(top_auc = max(AUC)) %>%
  filter(top_auc == max(top_auc)) %>%
  arrange(-top_auc)

# Plot ROC curves
cutpoints_unnest_shannon %>%
  arrange(desc(AUC)) %>%
  group_by(pos_class) %>%
  ungroup() %>%
  unnest(roc_curve) %>%
  arrange(desc(AUC)) %>%
  mutate(auc_label = paste0("AUC = ", round(AUC, 5))) %>%
  ggplot(aes(x = fpr, y = tpr)) +
  geom_line() +
  geom_text(aes(label = auc_label, x = 0.6, y = 0.2)) +
  geom_text(aes(
    label = round(optimal_cutpoint, 3),
    y = 0.8,
    x = 0.2
  )) +
  facet_wrap(~pos_class)

cutpoints_results_var_slct_shannon <- cutpoints_df_shannon %>%
  mutate(
    shannon_class = ifelse(
      Shannon >= cutpoints_unnest_summary_shannon$optimal_cutpoint,
      paste0(
        "High Diversity (Shannon > ",
        round(cutpoints_unnest_summary_shannon$optimal_cutpoint, 2),
        ")"
      ),
      paste0(
        "Low Diversity (Shannon < ",
        round(cutpoints_unnest_summary_shannon$optimal_cutpoint, 2),
        ")"
      )
    ),
    shannon_binary = ifelse(
      Shannon >= cutpoints_unnest_summary_shannon$optimal_cutpoint,
      1,
      0
    )
  )

# ROC curve for Shannon Diversity using training data
pROC_obj_shannon <- pROC::roc(
  cutpoints_results_var_slct_shannon$thirtyday_mortality_overall_class,
  cutpoints_results_var_slct_shannon$Shannon,
  smoothed = TRUE,
  ci = TRUE,
  plot = FALSE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

coordinates_shannon <-
  coords(pROC_obj_shannon,
    "best",
    ret = c("acc", "threshold", "sens", "spec", "ppv", "npv")
  )

pROC::roc(
  cutpoints_results_var_slct_shannon$thirtyday_mortality_overall_class,
  cutpoints_results_var_slct_shannon$Shannon,
  smoothed = TRUE,
  ci = TRUE,
  plot = TRUE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

text(
  paste("ACC:", round(coordinates_shannon$accuracy, 3) * 100, "%"),
  x = 0.5,
  y = 0.45,
  adj = 0
)
text(paste("PPV:", round(coordinates_shannon$ppv, 2)),
  x = 0.5,
  y = 0.41,
  adj = 0
)
text(paste("NPV:", round(coordinates_shannon$npv, 2)),
  x = 0.5,
  y = 0.37,
  adj = 0
)
text(paste("Threshold:", round(coordinates_shannon$threshold, 2)),
  x = 0.5,
  y = 0.33,
  adj = 0
)

roc_plot_shannon <- grDevices::recordPlot()

cairo_pdf(
  "./Results/ROC_curve_cutpoint_shannon_30_Day_Mortality_train.pdf",
  width = 8,
  height = 6
)
pROC::roc(
  cutpoints_results_var_slct_shannon$thirtyday_mortality_overall_class,
  cutpoints_results_var_slct_shannon$Shannon,
  smoothed = TRUE,
  ci = TRUE,
  plot = TRUE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)
text(
  paste("ACC:", round(coordinates_shannon$accuracy, 3) * 100, "%"),
  x = 0.5,
  y = 0.45,
  adj = 0
)
text(paste("PPV:", round(coordinates_shannon$ppv, 2)),
  x = 0.5,
  y = 0.41,
  adj = 0
)
text(paste("NPV:", round(coordinates_shannon$npv, 2)),
  x = 0.5,
  y = 0.37,
  adj = 0
)
text(paste("Threshold:", round(coordinates_shannon$threshold, 2)),
  x = 0.5,
  y = 0.33,
  adj = 0
)
invisible(dev.off())

model_comps_df <- micu_nocovid_first_samps_omics_light %>%
  filter(metabolomicsID %in% micu_new_nocovid_oc$metabolomicsID) %>%
  group_by(unique_id, shotgunSeq_id, taxid) %>%
  dplyr::slice(1) %>%
  left_join(taxdmp %>% mutate(taxid = as.character(taxid))) %>%
  mutate(Species = paste(Kingdom, Phylum, Class, Order, Family, Genus, Species, sep = "|")) %>%
  filter(grepl(pattern = "Enterococcus|Enterobacterales", x = Species)) %>%
  mutate(organism = case_when(
    grepl(pattern = "Enterococcus", x = Species) ~ "Enterococcus",
    grepl(pattern = "Enterobacterales", x = Species) ~ "Enterobacterales",
    TRUE ~ NA
  )) %>%
  drop_na(organism) %>%
  select(unique_id, thirtyday_mortality_overall, organism, pctseqs) %>%
  group_by(unique_id, thirtyday_mortality_overall, organism) %>%
  summarise(pctseqs = sum(pctseqs)) %>%
  pivot_wider(names_from = "organism", values_from = "pctseqs") %>% 
  left_join(micu_new_nocovid_oc %>% select(unique_id, shotgunSeq_id, metabolomicsID)) %>%
  select(unique_id, shotgunSeq_id, metabolomicsID, Enterococcus, Enterobacterales) %>%
  left_join(alpha_shannon) %>%
  left_join(tableone_nocovid_df_filt %>% select(unique_id, thirtyday_mortality_overall)) %>%
  left_join(km_nocovid_final)

# MD Score
pROC_obj_mds <- pROC::roc(
  model_comps_df$thirtyday_mortality_overall,
  model_comps_df$md_score,
  smoothed = TRUE,
  ci = TRUE,
  plot = FALSE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

coordinates_mds <-
  coords(pROC_obj_mds,
    "best",
    ret = c("acc", "threshold", "sens", "spec", "ppv", "npv")
  )

pROC::roc(
  model_comps_df$thirtyday_mortality_overall,
  model_comps_df$md_score,
  smoothed = TRUE,
  ci = TRUE,
  plot = TRUE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

text(
  paste("ACC:", round(coordinates_mds$accuracy, 3) * 100, "%"),
  x = 0.5,
  y = 0.45,
  adj = 0
)
text(paste("PPV:", round(coordinates_mds$ppv, 2)),
  x = 0.5,
  y = 0.41,
  adj = 0
)
text(paste("NPV:", round(coordinates_mds$npv, 2)),
  x = 0.5,
  y = 0.37,
  adj = 0
)
text(paste("Threshold:", round(coordinates_mds$threshold, 2)),
  x = 0.5,
  y = 0.33,
  adj = 0
)

# MD Score Model Metrics
coordinates_mds_df <- model_comps_df %>%
  bind_cols(coordinates_mds$threshold[1]) %>% 
  dplyr::rename(threshold = ...19) %>%
  mutate(prediction = ifelse(md_score > threshold, "Non-Survivor", "Survivor"))

caret::confusionMatrix(table(
  factor(
    coordinates_mds_df$prediction,
    levels = c("Survivor", "Non-Survivor")
  ),
  factor(
    coordinates_mds_df$thirtyday_mortality_overall,
    levels = c("Survivor", "Non-Survivor")
  )
))

# Confusion Matrix and Statistics
# 
#               
#                Survivor Non-Survivor
#   Survivor           86           11
#   Non-Survivor       16           34
#                                           
#                Accuracy : 0.8163          
#                  95% CI : (0.7441, 0.8753)

# Build df of MDS and Shannon
mds_shannon <- micu_new_nocovid_oc %>%
  select(shotgunSeq_id, metabolomicsID) %>%
  right_join(km_nocovid_final %>% select(metabolomicsID, thirtyday_mortality_overall, md_score)) %>%
  left_join(cutpoints_results_var_slct_shannon) %>%
  select(Shannon, md_score) %>%
  mutate(
    outlier = ifelse(
      Shannon < coordinates_shannon$threshold &
        md_score < coordinates_mds$threshold |
        Shannon > coordinates_shannon$threshold &
          md_score > coordinates_mds$threshold,
      "Outlier",
      "Not Outlier"
    )
  )

# Correlation plot of MDS with Shannon
gg_mds_shannon_scatter <-
  ggscatter(
    mds_shannon,
    y = "Shannon",
    x = "md_score",
    size = 3,
    color = "outlier",
    alpha = 0.2,
    palette = "lancet",
    add = "reg.line",
    add.params = list(color = "black"),
    conf.int = TRUE
  ) +
  stat_cor(method = "spearman") +
  geom_vline(xintercept = coordinates_mds$threshold, linetype = "longdash") +
  geom_hline(yintercept = coordinates_shannon$threshold, linetype = "longdash") +
  ylab("\U03B1-Diversity\n") +
  xlab("\nMD Score") +
  guides(color = guide_legend("Outlier"))

gg_mds_shannon_scatter

cairo_pdf(
  "./Results/MDS_Shannon_Correlation_train.pdf",
  height = 6,
  width = 8
)
gg_mds_shannon_scatter
invisible(dev.off())
```

#### ROC and Correlation Analyses: Enterococcus/Enterobacterales Relative Abundance and MDS
```{r, md-score-ecoc-relative-abundance-model-metrics}
# Enterococcus Relative Abundance
pROC_obj_ecoc <- pROC::roc(
  model_comps_df$thirtyday_mortality_overall,
  model_comps_df$Enterococcus,
  smoothed = TRUE,
  ci = TRUE,
  plot = FALSE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

coordinates_ecoc <-
  pROC::coords(pROC_obj_ecoc,
    "best",
    ret = c("acc", "threshold", "sens", "spec", "ppv", "npv")
  )

pROC::roc(
  model_comps_df$thirtyday_mortality_overall,
  model_comps_df$Enterococcus,
  smoothed = TRUE,
  ci = TRUE,
  plot = TRUE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

text(
  paste("ACC:", round(coordinates_ecoc$accuracy, 3) * 100, "%"),
  x = 0.5,
  y = 0.45,
  adj = 0
)
text(paste("PPV:", round(coordinates_ecoc$ppv, 2)),
  x = 0.5,
  y = 0.41,
  adj = 0
)
text(paste("NPV:", round(coordinates_ecoc$npv, 2)),
  x = 0.5,
  y = 0.37,
  adj = 0
)
text(paste("Threshold:", round(coordinates_ecoc$threshold, 2)),
  x = 0.5,
  y = 0.33,
  adj = 0
)

# Enterococcus Relative Abundance Model Metrics
coordinates_mmp_ecoc <- model_comps_df %>%
  bind_cols(coordinates_ecoc$threshold) %>%
  dplyr::rename(threshold = ...19) %>%
  mutate(prediction = ifelse(Enterococcus >= threshold, "Non-Survivor", "Survivor"))

caret::confusionMatrix(table(
  factor(
    coordinates_mmp_ecoc$prediction,
    levels = c("Survivor", "Non-Survivor")
  ),
  factor(
    coordinates_mmp_ecoc$thirtyday_mortality_overall,
    levels = c("Survivor", "Non-Survivor")
  )
))

# Confusion Matrix and Statistics

 #               Survivor Non-Survivor
 #  Survivor           52            6
 #  Non-Survivor       50           39
 #                                          
 #               Accuracy : 0.619           
 #                 95% CI : (0.5354, 0.6978)
 #    No Information Rate : 0.6939          
 #    P-Value [Acc > NIR] : 0.9786          
 #                                          
 #                  Kappa : 0.2957          
 #                                          
 # Mcnemar's Test P-Value : 9.132e-09       
 #                                          
 #            Sensitivity : 0.5098          
 #            Specificity : 0.8667          

# Build df of Shannon and Enterococcus
mds_ecoc <- model_comps_df %>%
  select(md_score, Enterococcus) %>%
  mutate(
    outlier = ifelse(
      md_score < coordinates_mds$threshold & Enterococcus >= 0.199,
      "Outlier",
      "Not Outlier"
    )
  )

# Correlation plot of MDS with Enterococcus
gg_mds_ecoc_scatter <-
  ggscatter(
    mds_ecoc,
    y = "Enterococcus",
    x = "md_score",
    size = 3,
    color = "outlier",
    alpha = 0.2,
    palette = "lancet",
    add = "reg.line",
    add.params = list(color = "black"),
    conf.int = TRUE
  ) +
  stat_cor(method = "spearman") +
  geom_hline(yintercept = 0.30, linetype = "longdash") +
  geom_vline(xintercept = coordinates_mds$threshold, linetype = "longdash") +
  xlab("\nMD Score") +
  ylab("Enterococcus Relative Abundance (%)\n") +
  guides(color = guide_legend("Outlier"))

gg_mds_ecoc_scatter

cairo_pdf(
  "./Results/MDS_Enterococcus_Correlation_train.pdf",
  height = 6,
  width = 8
)
gg_mds_ecoc_scatter
invisible(dev.off())

# Number of Ecoc Expansions/High MDS and Ecoc Expansions/Low MDS
model_comps_df %>%
  select(md_score, Enterococcus) %>%
  mutate(
    measure = case_when(
      Enterococcus >= 0.199 &
        md_score < coordinates_mds$threshold ~ "Expan_LMDS",
      Enterococcus >= 0.199 &
        md_score >= coordinates_mds$threshold ~ "Expan_HMDS",
      Enterococcus < 0.199 &
        md_score < coordinates_mds$threshold ~ "NoExpan_LMDS",
      Enterococcus < 0.199 &
        md_score >= coordinates_mds$threshold ~ "NoExpan_HMDS"
    )
  ) %>%
  group_by(measure) %>%
  tally() %>%
  dplyr::rename(count = n) %>%
  mutate(
    total = sum(count),
    percent = (count / total) * 100
  ) %>%
  write.csv(.,
    "./Results/Enterococcus_Expansion_MD_Score_train.csv",
    row.names = FALSE
  )

# Enterobacterales Relative Abundance
pROC_obj_ebac <- pROC::roc(
  model_comps_df$thirtyday_mortality_overall,
  model_comps_df$Enterobacterales,
  smoothed = TRUE,
  ci = TRUE,
  plot = FALSE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

coordinates_ebac <-
  pROC::coords(pROC_obj_ebac,
    "best",
    ret = c("acc", "threshold", "sens", "spec", "ppv", "npv")
  )

pROC::roc(
  model_comps_df$thirtyday_mortality_overall,
  model_comps_df$Enterobacterales,
  smoothed = TRUE,
  ci = TRUE,
  plot = TRUE,
  auc.polygon = TRUE,
  best.method = TRUE,
  print.auc = TRUE,
  print.auc.col = "black",
  col = "#2F472F",
  auc.polygon.border = "black",
  auc.polygon.col = "gray65",
  print.thres.best.method = "youden"
)

text(
  paste("ACC:", round(coordinates_ebac$accuracy, 3) * 100, "%"),
  x = 0.5,
  y = 0.45,
  adj = 0
)
text(paste("PPV:", round(coordinates_ebac$ppv, 2)),
  x = 0.5,
  y = 0.41,
  adj = 0
)
text(paste("NPV:", round(coordinates_ebac$npv, 2)),
  x = 0.5,
  y = 0.37,
  adj = 0
)
text(paste("Threshold:", round(coordinates_ebac$threshold, 2)),
  x = 0.5,
  y = 0.33,
  adj = 0
)

# Enterobacterales Relative Abundance Model Metrics
coordinates_mmp_ebacc <- model_comps_df %>%
  bind_cols(coordinates_ebac$threshold) %>%
  dplyr::rename(threshold = ...19) %>%
  mutate(prediction = ifelse(Enterobacterales >= threshold, "Non-Survivor", "Survivor"))

caret::confusionMatrix(table(
  factor(
    coordinates_mmp_ecoc$prediction,
    levels = c("Survivor", "Non-Survivor")
  ),
  factor(
    coordinates_mmp_ecoc$thirtyday_mortality_overall,
    levels = c("Survivor", "Non-Survivor")
  )
))

# Confusion Matrix and Statistics
# 
#               
#                Survivor Non-Survivor
#   Survivor           52            6
#   Non-Survivor       50           39
#                                           
#                Accuracy : 0.619           
#                  95% CI : (0.5354, 0.6978)
#     No Information Rate : 0.6939          
#     P-Value [Acc > NIR] : 0.9786          
#                                           
#                   Kappa : 0.2957          
#                                           
#  Mcnemar's Test P-Value : 9.132e-09       
#                                           
#             Sensitivity : 0.5098          
#             Specificity : 0.8667 

# Build df of Shannon and Enterobacterales
mds_ebac <- model_comps_df %>%
  select(md_score, Enterobacterales) %>%
  mutate(
    outlier = ifelse(
      md_score < coordinates_mds$threshold & Enterobacterales >= 0.025,
      "Outlier",
      "Not Outlier"
    )
  )

# Correlation plot of MDS with Enterobacterales
gg_mds_ebac_scatter <-
  ggscatter(
    mds_ebac,
    y = "Enterobacterales",
    x = "md_score",
    size = 3,
    color = "outlier",
    alpha = 0.2,
    palette = "lancet",
    add = "reg.line",
    add.params = list(color = "black"),
    conf.int = TRUE
  ) +
  stat_cor(method = "spearman") +
  geom_hline(yintercept = 0.30, linetype = "longdash") +
  geom_vline(xintercept = coordinates_mds$threshold, linetype = "longdash") +
  xlab("\nMD Score") +
  ylab("Enterobacterales Relative Abundance (%)\n") +
  guides(color = guide_legend("Outlier"))

gg_mds_ebac_scatter

cairo_pdf(
  "./Results/MDS_Enterobacterales_Correlation_train.pdf",
  height = 6,
  width = 8
)
gg_mds_ebac_scatter
invisible(dev.off())

# Number of Ebac Expansions/High MDS and Ebac Expansions/Low MDS
model_comps_df %>%
  select(md_score, Enterobacterales) %>%
  mutate(
    measure = case_when(
      Enterobacterales >= 0.025 &
        md_score < coordinates_mds$threshold ~ "Expan_LMDS",
      Enterobacterales >= 0.025 &
        md_score >= coordinates_mds$threshold ~ "Expan_HMDS",
      Enterobacterales < 0.025 &
        md_score < coordinates_mds$threshold ~ "NoExpan_LMDS",
      Enterobacterales < 0.025 &
        md_score >= coordinates_mds$threshold ~ "NoExpan_HMDS"
    )
  ) %>%
  group_by(measure) %>%
  tally() %>%
  dplyr::rename(count = n) %>%
  mutate(
    total = sum(count),
    percent = (count / total) * 100
  ) %>%
  write.csv(.,
    "./Results/Enterobacterales_Expansion_MD_Score_train.csv",
    row.names = FALSE
  )
```

#### Kaplan-Meier Survival Analyses: MDS (5.5), Enterococcus (19.9%), Enterobacterales (2.5%), Shannon (2.16)
```{r, kaplan-meier}
km_nocovid <- micu_new_nocovid_oc %>%
  select(
    unique_id,
    sampleid,
    metabolomicsID,
    days_until_death_overall,
    censoring_thirtyday_mortality_overall,
    thirtyday_mortality_overall
  ) %>%
  ungroup() %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>% # Non-Survivor is 1
  left_join(km_nocovid_final %>% select(metabolomicsID, thirtyday_mortality_overall, grouped_md_score)) %>%
  left_join(model_comps_df %>%
      mutate(enterococcus_domination_threshold = ifelse(Enterococcus >= 0.199, 1, 0),
             enterobacterales_domination_threshold = ifelse(Enterobacterales >= 0.025, 1, 0)) %>%
      select(unique_id, enterococcus_domination_threshold, enterobacterales_domination_threshold)
  ) %>%
  left_join(
    alpha_shannon %>%
      left_join(
        micu_new_nocovid_oc %>%
          ungroup() %>%
          select(unique_id, shotgunSeq_id) %>%
          distinct(shotgunSeq_id, .keep_all = TRUE),
        by = "shotgunSeq_id"
      )
  ) %>%
  mutate(
    shannon_class = ifelse(
      Shannon >= coords(
        pROC_obj_shannon,
        "best",
        ret = c("threshold", "sens", "spec", "ppv", "npv")
      )[1][[1]],
      paste0(
        "High Diversity (Shannon > ",
        round(cutpoints_unnest_summary_shannon$optimal_cutpoint, 2),
        ")"
      ),
      paste0(
        "Low Diversity (Shannon < ",
        round(cutpoints_unnest_summary_shannon$optimal_cutpoint, 2),
        ")"
      )
    ),
    enterococcus_domination_threshold = ifelse(
      enterococcus_domination_threshold == 1,
      "Enterococcous Domination",
      "No Enterococcus Domination"
    ),
    enterococcus_domination_threshold = factor(
      enterococcus_domination_threshold,
      levels = c("Enterococcous Domination", "No Enterococcus Domination")
    ),
    enterobacterales_domination_threshold = ifelse(
      enterobacterales_domination_threshold == 1,
      "Enterobacterales Domination",
      "No Enterobacterales Domination"
    ),
    enterobacterales_domination_threshold = factor(
      enterobacterales_domination_threshold,
      levels = c("Enterobacterales Domination", "No Enterobacterales Domination")
    )
  )

# KM Curves: MD Score
set.seed(123)
surv_object <-
  Surv(
    time = km_nocovid$surv_days,
    event = km_nocovid$thirtyday_mortality_overall_class
  )

fit1 <- survfit(surv_object ~ grouped_md_score, data = km_nocovid)

ggs <- ggsurvplot(
  fit1,
  data = km_nocovid,
  size = 1,
  palette = c("#C45258", "#2F4858"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  legend.labs = c("High MD Score", "Low MD Score")
)

# Change table axis labels
ggs$table <-
  ggs$table + labs(x = NULL, y = NULL) + theme(plot.title = eb()) # risk table

ggs

pdf(
  "./Results/kaplan_meier_roc_loop_30_Day_Mortality_train.pdf",
  height = 4,
  width = 6,
  onefile = FALSE
)
ggs
invisible(dev.off())

# Restricted Mean Survival Time
rmst_mds <-
survRM2::rmst2(
  km_nocovid$surv_days,
  km_nocovid$thirtyday_mortality_overall_class,
  factor(
    km_nocovid$grouped_md_score,
    levels = c("Low Score", "High Score"),
    labels = c(1, 0)                      # Low Score = 1, High Score = 0, due to area = Low Score - High Score
  ),
  tau = 30
)

plot(rmst_mds, xlab = "Days", ylab = "Survival Probability")

# KM Curves: Enterococcus Domination (>= 0.199) #coordinates_ecoc$threshold)
set.seed(123)
surv_object2 <-
  Surv(
    time = km_nocovid$surv_days,
    event = km_nocovid$thirtyday_mortality_overall_class
  )

fit2 <-
  survfit(surv_object2 ~ enterococcus_domination_threshold, data = km_nocovid)

ggs_ecoc <- ggsurvplot(
  fit2,
  data = km_nocovid,
  size = 1,
  palette = c("#C4335F", "#047D6B"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  surv.median.line = "hv",
  legend.labs = c(
    paste0("Enterococcus Domination (>", round(0.199
    * 100, 2), "%)"),
    "No Domination"
  )
)

# Change table axis labels
ggs_ecoc$table <-
  ggs_ecoc$table + labs(x = NULL, y = NULL) + theme(plot.title = eb()) # risk table

ggs_ecoc

pdf(
  "./Results/kaplan_meier_enterococcus_30_Day_Mortality_train.pdf",
  height = 4,
  width = 6,
  onefile = FALSE
)
ggs_ecoc
invisible(dev.off())

# KM Curves: Enterobacterales Domination (>= 0.025) #coordinates_ebac$threshold)
set.seed(123)
surv_object3 <-
  Surv(
    time = km_nocovid$surv_days,
    event = km_nocovid$thirtyday_mortality_overall_class
  )

fit3 <-
  survfit(surv_object2 ~ enterobacterales_domination_threshold, data = km_nocovid)

ggs_ebac <- ggsurvplot(
  fit3,
  data = km_nocovid,
  size = 1,
  palette = c("#C4335F", "#047D6B"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  surv.median.line = "hv",
  legend.labs = c(
    paste0("Enterobacterales Domination (>", round(0.025
    * 100, 2), "%)"),
    "No Domination"
  )
)

# Change table axis labels
ggs_ebac$table <-
  ggs_ebac$table + labs(x = NULL, y = NULL) + theme(plot.title = eb()) # risk table

ggs_ebac

pdf(
  "./Results/kaplan_meier_enterobacterales_30_Day_Mortality_train.pdf",
  height = 4,
  width = 6,
  onefile = FALSE
)
ggs_ebac
invisible(dev.off())

# KM Curves: Shannon Diversity (>cutpoints_unnest_summary_shannon$optimal_cutpoint)
set.seed(123)
surv_object4 <-
  Surv(
    time = km_nocovid$surv_days,
    event = km_nocovid$thirtyday_mortality_overall_class
  )

fit4 <- survfit(surv_object4 ~ shannon_class, data = km_nocovid)

ggs_shannon <- ggsurvplot(
  fit4,
  data = km_nocovid,
  size = 1,
  palette = c("#107b1c","#C45258"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  surv.median.line = "hv",
  legend.labs = c(
    paste0("High Diversity (Shannon >", round(
      coords(
        pROC_obj_shannon,
        "best",
        ret = c("threshold", "sens", "spec", "ppv", "npv")
      )[1][[1]], 2
    ), ")"),
    paste0("Low Diversity (Shannon <=", round(
      coords(
        pROC_obj_shannon,
        "best",
        ret = c("threshold", "sens", "spec", "ppv", "npv")
      )[1][[1]], 2
    ), ")")
  )
)

# Change table axis labels
ggs_shannon$table <-
  ggs_shannon$table + labs(x = NULL, y = NULL) + 
  theme(plot.title = eb()) # risk table

ggs_shannon

pdf(
  "./cox/kaplan_meier_shannon_30_Day_Mortality_train.pdf",
  height = 4,
  width = 6,
  onefile = FALSE
)
ggs_shannon
invisible(dev.off())
```

#### Cox Proportional Hazards Regression Analyses: MDS, Enterococcus, Enterobacterales, Shannon
```{r, cox-regression}
# Variables labels
cox_df <- tableone_nocovid_df_filt %>%
  labelled::remove_labels() %>%
  janitor::clean_names() %>%
  mutate(
    race_factor = as.character(race_factor),
    race_factor = ifelse(
      race_factor %in% c("Asian", "More than one race", "White, Hispanic"),
      "Other",
      race_factor
    )
  ) %>%
  left_join(
    micu_nocovid_first_samps_omics_light %>%
      group_by(metabolomicsID) %>%
      slice(1) %>%
      select(unique_id, metabolomicsID)
  ) %>%
  left_join(km_nocovid_final %>% select(metabolomicsID, md_score)) %>%
  mutate(grouped_md_score = ifelse(
    md_score >= coordinates_mds$threshold,
    "High Score",
    "Low Score"
  )) %>%
  right_join(
    micu_new_nocovid_oc %>% select(
      unique_id,
      days_until_death_overall,
      censoring_thirtyday_mortality_overall,
      thirtyday_mortality_overall
    )
  ) %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>%
  group_by(metabolomicsID) %>%
  dplyr::slice(1) %>%
  left_join(
    alpha_shannon %>%
      left_join(
        micu_new_nocovid_oc %>%
          ungroup() %>%
          select(unique_id, shotgunSeq_id) %>%
          distinct(shotgunSeq_id, .keep_all = TRUE),
        by = "shotgunSeq_id"
      )
  ) %>%
  left_join(
    model_comps_df %>%
      mutate(enterococcus_domination_threshold = ifelse(Enterococcus >= 0.199, 1, 0),
             enterobacterales_domination_threshold = ifelse(Enterobacterales >= 0.025, 1, 0)) %>%
      select(unique_id, enterococcus_domination_threshold, enterobacterales_domination_threshold)
  ) %>%
  mutate(
    shannon_class = ifelse(
      Shannon >= coords(
        pROC_obj_shannon,
        "best",
        ret = c("threshold", "sens", "spec", "ppv", "npv")
      )[1][[1]],
      paste0(
        "High Diversity (Shannon > ",
        round(cutpoints_unnest_summary_shannon$optimal_cutpoint, 2),
        ")"
      ),
      paste0(
        "Low Diversity (Shannon < ",
        round(cutpoints_unnest_summary_shannon$optimal_cutpoint, 2),
        ")"
      )
    ),
    enterococcus_domination_threshold = ifelse(
      enterococcus_domination_threshold == 1,
      "Enterococcous Domination",
      "No Enterococcus Domination"
    ),
    enterococcus_domination_threshold = factor(
      enterococcus_domination_threshold,
      levels = c("Enterococcous Domination", "No Enterococcus Domination")
    ),
    enterobacterales_domination_threshold = ifelse(
      enterobacterales_domination_threshold == 1,
      "Enterobacterales Domination",
      "No Enterobacterales Domination"
    ),
    enterobacterales_domination_threshold = factor(
      enterobacterales_domination_threshold,
      levels = c("Enterobacterales Domination", "No Enterobacterales Domination")
    )
  ) %>%
  dplyr::rename(`Charlson Comorbidity Index` = cci_total_sc) %>%
  mutate(diet = ifelse(diet == "1", "Diet", "NPO")) %>% 
  dplyr::rename(
    `Sex` = "sex_factor",
    `Age` = "age",
    `Acute respiratory distress syndrome` = "ards_factor",
    `Sepsis` = "sepsis_factor",
    `SOFA Score` = "sofa_score_total",
    `Race` = "race_factor",
    `Time to stool sample` = "day_collected",
    `Diet` = "diet",
    `MDS` = "md_score",
    `Enterococcus Domination` = "enterococcus_domination_threshold",
    `Enterobacterales Domination` = "enterobacterales_domination_threshold",
    `Shannon Diversity` = "Shannon"
  )


reset_gtsummary_theme()

coxauc <-
  coxph(
    Surv(cox_df$surv_days, cox_df$thirtyday_mortality_overall_class) ~
      `Charlson Comorbidity Index` +
      `SOFA Score` +
      `MDS`,
    data = cox_df
  ) %>%
  gtsummary::tbl_regression(
    exp = TRUE,
    pvalue_fun = function(x) {
      if_else(is.na(x), NA_character_, if_else(
        x < 0.001,
        format(x,
          digits = 3, scientific = TRUE
        ),
        format(round(x, 3),
          scientific = F
        )
      ))
    }
  ) %>%
  gtsummary::modify_footnote(everything() ~ NA, abbreviation = TRUE)


coxauc %>%
  gtsummary::modify_caption("**Cox Proportional Hazards Regression**")

# In case you get an error: "Error in s$close() : attempt to apply non-function", run this code below:
# f <- chromote::default_chromote_object() #get the f object
# f$close()

gt::gtsave(gtsummary::as_gt(coxauc), file = "cox/cox_model_SOFA_30_Day_Mortality_roc_loop_train.png")

cox_df <- cox_df |> 
  mutate(`Enterococcus Domination` =
           factor(`Enterococcus Domination`, levels = c("No Enterococcus Domination", "Enterococcous Domination")))

# Enterococcus Domination: as Yes/No
coxauc_ecoc <-
  coxph(
    Surv(cox_df$surv_days, cox_df$thirtyday_mortality_overall_class) ~
      `Charlson Comorbidity Index` +
      `SOFA Score` +
      `Enterococcus Domination`,
    data = cox_df
  ) %>%
  gtsummary::tbl_regression(
    exp = TRUE,
    pvalue_fun = function(x) {
      if_else(is.na(x), NA_character_, if_else(
        x < 0.001,
        format(x,
          digits = 3, scientific = TRUE
        ),
        format(round(x, 3),
          scientific = F
        )
      ))
    }
  ) %>%
  gtsummary::modify_footnote(everything() ~ NA, abbreviation = TRUE)

coxauc_ecoc %>%
  gtsummary::modify_caption("**Cox Proportional Hazards Regression**")

gt::gtsave(gtsummary::as_gt(coxauc_ecoc), file = "./Results/cox_model_SOFA_Enterococcus_30_Day_Mortality_train.png")

# Enterococcus Domination
coxauc_ebac <-
  coxph(
    Surv(cox_df$surv_days, cox_df$thirtyday_mortality_overall_class) ~
      `Charlson Comorbidity Index` +
      `SOFA Score` +
      `Enterobacterales Domination`,
    data = cox_df
  ) %>%
  tbl_regression(
    exp = TRUE,
    pvalue_fun = function(x) {
      if_else(is.na(x), NA_character_, if_else(
        x < 0.001,
        format(x,
          digits = 3, scientific = TRUE
        ),
        format(round(x, 3),
          scientific = F
        )
      ))
    }
  ) %>%
  modify_footnote(everything() ~ NA, abbreviation = TRUE)

coxauc_ebac %>%
  gtsummary::modify_caption("**Cox Proportional Hazards Regression**")

gt::gtsave(gtsummary::as_gt(coxauc_ebac), file = "cox/cox_model_SOFA_Enterobacterales_30_Day_Mortality_train.png")

# Shannon Diversity
coxauc_shannon <-
  coxph(
    Surv(cox_df$surv_days, cox_df$thirtyday_mortality_overall_class) ~
      `Charlson Comorbidity Index` +
      `SOFA Score` +
      `Shannon Diversity`,
    data = cox_df
  ) %>%
  tbl_regression(
    exp = TRUE,
    pvalue_fun = function(x) {
      if_else(is.na(x), NA_character_, if_else(
        x < 0.001,
        format(x,
          digits = 3, scientific = TRUE
        ),
        format(round(x, 3),
          scientific = F
        )
      ))
    }
  ) %>%
  modify_footnote(everything() ~ NA, abbreviation = TRUE)


coxauc_shannon %>%
  gtsummary::modify_caption("**Cox Proportional Hazards Regression**")

gt::gtsave(gtsummary::as_gt(coxauc_shannon), file = "cox/cox_model_SOFA_Shannon_30_Day_Mortality_train.png")
```

#### Beta Diversity
```{r, beta-diversity}
thirtyday_mortality_overall_vector <-
  t_metaphlan_micu_nocovid_mat %>%
  rownames_to_column(var = "shotgunSeq_id") %>%
  select(shotgunSeq_id) %>%
  left_join(micu_new_nocovid_oc %>%
    select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  column_to_rownames(var = "shotgunSeq_id") %>%
  pull(thirtyday_mortality_overall)

beta_dist <-
  vegdist(t_metaphlan_micu_nocovid_mat, index = "bray-curtis")

mds <-
  metaMDS(
    beta_dist,
    k = 3,
    distance = "bray-curtis",
    trymax = 500,
    wascores = TRUE
  )

mds_data <- as.data.frame(mds$points)

# Shepards test/goodness of fit
goodness(mds) # Produces a results of test statistics for goodness of fit for each point

stressplot(mds) # Produces a Shepards diagram

# Stats: Homogeneity of dispersion test
dispersion <-
  permutest(betadisper(beta_dist, thirtyday_mortality_overall_vector)) # No significant difference in dispersion between Survivor and Non-Survivor

dispersion_pval <- dispersion$tab$`Pr(>F)`[1]

# Stats: PERMANOVA
set.seed(123)
mds_stats <-
  adonis2(
    beta_dist ~ thirtyday_mortality_overall_vector,
    method = "bray-curtis",
    permutations = 999
  )

mds_pval <- mds_stats$`Pr(>F)`[1]

# Stats: Pairwise analysis
pair_mod <-
  pairwise.adonis(beta_dist, factors = thirtyday_mortality_overall_vector, p.adjust.m = "BH")
pair_mod

mds_data2 <- mds_data %>%
  rownames_to_column(var = "shotgunSeq_id") %>%
  left_join(micu_new_nocovid_oc %>%
    select(shotgunSeq_id, thirtyday_mortality_overall))

ggplot_mds <-
  ggplot(
    mds_data2,
    aes(
      x = MDS1,
      y = MDS2,
      color = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  stat_ellipse(
    level = 0.1,
    geom = "polygon",
    alpha = 0.35,
    type = "euclid"
  ) +
  geom_point(alpha = 0.65, size = 10) +
  theme_bw() +
  theme(
    axis.title = et(color = "black", size = 72),
    axis.text = et(color = "black", size = 60),
    # plot.subtitle = et(color = "black", size = 79),
    panel.grid.minor = eb(),
    panel.grid.major = eb(),
    # legend.position = "none",
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5 
    )
  ) +
  annotate(
    "text",
    x = -0.4,
    y = 1.5,
    hjust = 0,
    size = 18,
    label = paste0(
      "BetaDisper = ", dispersion_pval, "\n",
      "PERMANOVA, p = ", mds_pval
    )
  ) +
  labs(
    y = "MDS2",
    x = "MDS1"
  ) +
  ggsci::scale_color_lancet() +
  ggsci::scale_fill_lancet() +
  guides(
    fill = guide_legend("Outcome"),
    color = guide_legend("Outcome")
  ) +
  coord_equal(
    ylim = c(-1.2, 1.7),
    xlim = c(-1.2, 1.7)
  )

ggplot_mds

ggsave(
  plot = ggplot_mds,
  filename = "./cox/Beta_Diversity_BrayCurtis_train.pdf",
  height = 14,
  width = 14,
  units = "in"
)
```

#### Qual Metabolites UMAP
```{r, qual-metabolomics-umap}
umap_metab_qual <- micu_new_nocovid_oc %>%
  left_join(metab_qual_imp_tot) %>%
  group_by(compound) %>%
  mutate(n = sum(is.na(mvalue))) %>%
  ungroup() %>%
  mutate(p = length(unique(metabolomicsID)))

umap_metab_qual_mat <- umap_metab_qual %>%
  select(metabolomicsID, compound, mvalue) %>%
  group_by(compound) %>%
  mutate(zscore = (mvalue - mean(mvalue, na.rm = TRUE)) / sd(mvalue, na.rm = TRUE)) %>%
  pivot_wider(metabolomicsID,
    names_from = "compound",
    values_from = "zscore"
  ) %>%
  purrr::discard(~ all(is.nan(.))) %>%
  column_to_rownames(var = "metabolomicsID") %>%
  janitor::remove_constant(.)

custom_config <- umap.defaults
custom_config$n_neighbors <-
  as.integer(nrow(umap_metab_qual_mat) * 0.1)
custom_config$random_state <- 123
custom_config$metric <- "manhattan"
custom_config$n_epochs <- 1000
custom_config$min_dist <- 0.1

umap_metab_qual_mat2 <-
  umap(umap_metab_qual_mat, config = custom_config)

umap_metab_qual_plot <- umap_metab_qual_mat2$layout %>%
  as.data.frame() %>%
  mutate(metabolomicsID = row.names(.)) %>%
  left_join(
    umap_metab_qual %>%
      group_by(metabolomicsID, thirtyday_mortality_overall) %>%
      dplyr::slice(1) %>%
      select(metabolomicsID, thirtyday_mortality_overall)
  ) %>%
  ggplot(aes(x = V1, y = V2, color = thirtyday_mortality_overall)) +
  geom_point(alpha = 0.65, size = 3.25) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title = et(color = "black", size = 14),
    axis.text = et(color = "black", size = 12),
    legend.title = et(color = "black", size = 14),
    legend.text = et(color = "black", size = 12)
  ) +
  ggtitle(
    paste0(
      "Qualitative Metabolomics: UMAP \nSurvivor vs Non-Survivor \n",
      "n = ",
      nrow(umap_metab_qual_mat),
      "\n",
      custom_config$n_neighbors,
      " Neighbors"
    )
  ) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  guides(
    color = guide_legend(title = "Outcome"),
    fill = guide_legend(title = "Outcome")
  ) +
  ggsci::scale_color_lancet() +
  ggsci::scale_fill_lancet()

umap_metab_qual_plot

ggsave(
  plot = umap_metab_qual_plot,
  filename = "./Results/Qual_Metab_UMAP_30_Day_Mortality_train.pdf",
  height = 8,
  width = 10,
  units = "in"
)
```

#### Qual Metabolites PCA
```{r, qual-metabolomics-pca}
# PCA dataframe
pca_metab_qual_mat <- umap_metab_qual %>%
  select(metabolomicsID, thirtyday_mortality_overall, compound, mvalue) %>%
  group_by(compound) %>%
  mutate(zscore = (mvalue - mean(mvalue, na.rm = TRUE)) / sd(mvalue, na.rm = TRUE)) %>%
  pivot_wider(
    id_cols = c(metabolomicsID, thirtyday_mortality_overall),
    names_from = "compound",
    values_from = "zscore"
  ) %>%
  purrr::discard(~ all(is.nan(.))) %>%
  column_to_rownames(var = "metabolomicsID") %>%
  janitor::remove_constant()

# PCA on correlation matrix
pca_res <-
  prcomp(pca_metab_qual_mat[, -1], center = FALSE, scale = FALSE)

summary(pca_res)

# Scree plot
fviz_eig(pca_res, addlabels = TRUE)

# Biplot
fviz_pca_var(pca_res, col.var = "black", )

# Variable contribution plot
fviz_cos2(pca_res, choice = "var", axes = 1:2)

# Biplot plus Cos2 values
fviz_pca_var(
  pca_res,
  col.var = "cos2",
  gradient.cols = c("black", "orange", "green"),
  repel = TRUE
)

# Biplot plus contributions values
gg_pca_qual_vars <-
  fviz_pca_var(
    pca_res,
    col.var = "contrib",
    gradient.cols = c("white", "blue", "red"),
    ggtheme = theme_minimal(),
    repel = TRUE
  ) +
  theme(
    panel.grid = eb(),
    axis.text = et(size = 12, color = "black"),
    axis.title = et(size = 14, color = "black"),
    legend.title = et(size = 14, color = "black"),
    legend.text = et(size = 12, color = "black")
  ) +
  scale_x_continuous(expand = expansion(mult = c(0.25, 0.25))) +
  scale_y_continuous(expand = expansion(mult = c(0.25, 0.25))) +
  labs(color = "Contribution")

# Color individuals by outcome
gg_pca_qual_ind <-
  fviz_pca_ind(
    pca_res,
    label = "none",
    habillage = pca_metab_qual_mat$thirtyday_mortality_overall,
    addEllipses = TRUE,
    ellipse.level = 0.95,
    ggtheme = theme_minimal(),
  ) +
  theme(
    panel.grid = eb(),
    axis.text = et(size = 12, color = "black"),
    axis.title = et(size = 14, color = "black"),
    legend.title = et(size = 14, color = "black"),
    legend.text = et(size = 12, color = "black")
  ) +
  ggsci::scale_color_lancet() +
  ggsci::scale_fill_lancet()

pdf(
  file = "./Results/Qual_Metab_PCA_train.pdf",
  height = 8,
  width = 24
)
cowplot::plot_grid(gg_pca_qual_ind, gg_pca_qual_vars)
invisible(invisible(dev.off()))
```

#### Qual Metabolites Volcano
```{r, qual-metabolomics-volcano}
qual_log2fc <- umap_metab_qual %>%
  select(metabolomicsID, compound, mvalue, thirtyday_mortality_overall) %>%
  mutate(mvalue = ifelse(is.na(mvalue), 0, mvalue)) %>%
  group_by(compound) %>%
  filter(any(mvalue != 0)) %>%
  summarise(log2fc_val = log((
    mean(mvalue[thirtyday_mortality_overall == "Survivor"], na.rm = T) / mean(mvalue[thirtyday_mortality_overall == "Non-Survivor"], na.rm = T)
  ), base = 2)) %>%
  filter(compound != "pre-q1")

qual_pval <- umap_metab_qual %>%
  select(metabolomicsID, compound, mvalue, thirtyday_mortality_overall) %>%
  mutate(mvalue = ifelse(is.na(mvalue), 0, mvalue)) %>%
  group_by(compound) %>%
  filter(any(mvalue != 0)) %>%
  rstatix::wilcox_test(mvalue ~ thirtyday_mortality_overall) %>%
  rstatix::adjust_pvalue(method = "BH") %>%
  rstatix::add_significance("p.adj")

qual_tot <- left_join(qual_log2fc, qual_pval) %>%
  column_to_rownames(var = "compound")

write.csv(qual_tot, "./Results/volcano_list_train.csv")

# volcano label color
volcano_labcol <- qual_tot %>%
  filter(p.adj <= 0.1 & abs(log2fc_val) >= 0.75) %>%
  mutate(color = ifelse(
    log2fc_val > 0,
    ggsci::pal_lancet(palette = "lanonc")(2)[1],
    ggsci::pal_lancet(palette = "lanonc")(2)[2]
  ))

# Volcano Plot (adjusted)
set.seed(123)
volcano_adj <-
  EnhancedVolcano(
    qual_tot,
    lab = rownames(qual_tot),
    x = "log2fc_val",
    y = "p.adj",
    title = NULL,
    pCutoff = 0.1,
    FCcutoff = 0.75,
    pointSize = 6,
    labSize = 8,
    axisLabSize = 32,
    labCol = volcano_labcol$color,
    caption = NULL,
    colAlpha = 0.65,
    col = c("gray85", c("grey40", "grey10", "#F27DFA")),
    legendPosition = "bottom",
    legendLabels = c(
      expression(q > 0.1 * ";" ~ Log[2] ~ FC < "\u00B1" * 0.75),
      expression(q > 0.1 * ";" ~ Log[2] ~ FC >= "\u00B1" *
        0.75),
      expression(q <= 0.1 * ";" ~ Log[2] ~ FC < "\u00B1" *
        0.75),
      expression(q <= 0.1 * ";" ~ Log[2] ~ FC >= "\u00B1" *
        0.75)
    ),
    legendLabSize = 14,
    boxedLabels = TRUE,
    drawConnectors = TRUE,
    widthConnectors = 0.2,
    arrowheads = FALSE,
    gridlines.minor = FALSE,
    gridlines.major = FALSE,
    max.overlaps = Inf,
    min.segment.length = 0.5
  ) +
  theme(
    axis.text = et(color = "black"),
    legend.text = et(hjust = 0, size = 18),
    plot.margin = unit(c(0, 0, 0, 0), "cm")
  ) +
  labs(subtitle = NULL) +
  annotate(
    "segment",
    x = 0.8,
    xend = 2.5,
    y = 1.5,
    yend = 1.5,
    arrow = arrow(),
    size = 2,
    color = ggsci::pal_lancet(palette = "lanonc")(2)[1]
  ) +
  annotate(
    "text",
    x = 0.8,
    y = 1.6,
    hjust = 0,
    label = "Survivor",
    size = 12,
    color = ggsci::pal_lancet(palette = "lanonc")(2)[1]
  ) +
  annotate(
    "rect",
    xmin = 0.75,
    xmax = Inf,
    ymin = -log(0.1, base = 10),
    ymax = Inf,
    alpha = .1,
    fill = ggsci::pal_lancet(palette = "lanonc")(2)[1]
  ) +

  annotate(
    "segment",
    x = -0.8,
    xend = -2.5,
    y = 1.5,
    yend = 1.5,
    arrow = arrow(),
    size = 2,
    color = ggsci::pal_lancet(palette = "lanonc")(2)[2]
  ) +
  annotate(
    "text",
    x = -1.55,
    y = 1.6,
    hjust = 0.5,
    label = "Non-Survivor",
    size = 12,
    color = ggsci::pal_lancet(palette = "lanonc")(2)[2]
  ) +
  annotate(
    "rect",
    xmin = -0.75,
    xmax = -Inf,
    ymin = -log(0.1, base = 10),
    ymax = Inf,
    alpha = .1,
    fill = ggsci::pal_lancet(palette = "lanonc")(2)[2]
  ) +
  guides(
    color = guide_legend(nrow = 4),
    shape = guide_legend(nrow = 4)
  ) +
  scale_y_continuous(
    expand = expansion(mult = c(0, 0.05)),
    limits = c(0, 1.6),
    breaks = seq(0, 2, 0.5)
  )

volcano_adj

ggsave(
  plot = volcano_adj,
  filename = "./Results/Qual_Metab_Volcano_30_Day_Mortality_train.pdf",
  width = 24,
  height = 14
)
```

#### Metaphlan Relative Abundance, Alpha Diversity, and Enterococcus/Enterobacterales Relative Abundance
```{r, micu-mds-megaplot}
metaphlan_df2 <- t_metaphlan_micu_nocovid %>%
  left_join(taxdmp %>% mutate(taxid = as.character(taxid))) %>%
  drop_na(taxid) %>%
  arrange(Kingdom, Phylum, Class, Order, Family, Genus) %>%
  mutate(Genus = paste0(Phylum, "-", Order, "-", Family, "-", Genus)) %>%
  left_join(alpha_shannon) %>%
  group_by(shotgunSeq_id) %>%
  arrange(Genus) %>%
  mutate(
    cum.pct = cumsum(pctseqs),
    y.text = (cum.pct + c(0, cum.pct[-length(cum.pct)])) / 2
  ) %>%
  ungroup() %>%
  dplyr::select(-cum.pct)

metaphlan_pal <- getRdpPal(metaphlan_df2)

gg_metaphlan <- t_metaphlan_micu_nocovid %>%
  left_join(
    micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall) %>% mutate(
      thirtyday_mortality_overall = factor(
        thirtyday_mortality_overall,
        levels = c("Non-Survivor", "Survivor")
      )
    )
  ) %>%
  left_join(taxdmp %>% mutate(taxid = as.character(taxid))) %>%
  drop_na(taxid) %>%
  arrange(Kingdom, Phylum, Class, Order, Family, Genus) %>%
  mutate(Genus = paste0(Phylum, "-", Order, "-", Family, "-", Genus)) %>%
  left_join(alpha_shannon) %>%
  group_by(shotgunSeq_id) %>%
  mutate(
    cum.pct = cumsum(pctseqs),
    y.text = (cum.pct + c(0, cum.pct[-length(cum.pct)])) / 2
  ) %>%
  ungroup() %>%
  mutate(Genus = factor(Genus, levels = unique(Genus))) %>%
  group_by(shotgunSeq_id) %>%
  arrange(Genus) %>%
  ggplot(aes(x = reorder(shotgunSeq_id, Shannon), y = pctseqs)) +
  geom_bar(stat = "identity", aes(fill = Genus), width = 0.9) +
  scale_fill_manual(values = metaphlan_pal) +
  theme_bw() +
  theme(
    legend.position = "none",
    axis.text.x = eb(),
    axis.ticks.x = eb(),
    strip.text.x = et(angle = 0, size = 12),
    strip.background = eb(),
    axis.title.y = et(color = "black", size = 14),
    axis.text.y = et(color = "black", size = 12),
    panel.spacing = unit(0.5, "lines"),
    plot.margin = margin(
      t = 5,
      r = 5,
      b = 0,
      l = 5
    )
  ) +
  facet_grid(. ~ thirtyday_mortality_overall,
    scales = "free",
    space = "free_x"
  ) +
  scale_y_continuous(
    expand = expansion(mult = 0.005),
    labels = scales::percent_format(accuracy = 1)
  ) +
  scale_x_discrete(expand = expansion(add = 1)) +
  ylab("MetaPhlAn4 Relative Abundance\n") +
  xlab("")

gg_metaphlan

pdf(
  "./Results/Metaphlan_Relative_Abundance_train.pdf",
  height = 6,
  width = 12
)
gg_metaphlan
invisible(dev.off())

#### Alpha Diversity Plot Richness START ####
mat_filt <- t_metaphlan_micu_nocovid %>%
  pivot_wider(
    shotgunSeq_id,
    names_from = "taxid",
    values_from = "pctseqs",
    values_fill = 0
  ) %>%
  as.data.frame()

# Obtain stats for alpha diversity (Shannnon)
alpha_shannon_stats <-
  alpha_shannon %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  rstatix::wilcox_test(Shannon ~ thirtyday_mortality_overall)

pirate_colors <- rev(ggsci::pal_igv("default")(2))

set.seed(456)
gg_alpha_shannon <- alpha_shannon %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  mutate(
    thirtyday_mortality_overall = as.factor(thirtyday_mortality_overall),
    thirtyday_mortality_overall = factor(
      thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  ) %>%
  ggplot(
    .,
    aes(
      x = thirtyday_mortality_overall,
      y = Shannon,
      colour = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  geom_pirate(
    cis_params = list(fill = "white", alpha = 0.5),
    bars_params = list(alpha = 0.65),
    lines_params = list(size = 0.5),
    points_params = list(fill = "black", size = 3.5),
    jitter_width = 0.75,
    cis = TRUE,
    violins = FALSE
  ) +
  annotate(
    "text",
    x = 1.5,
    y = 5,
    label = paste0(
      "Wilcoxon, W = ",
      alpha_shannon_stats$statistic,
      ", p = ",
      alpha_shannon_stats$p
    ),
    size = 8
  ) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title.y = et(size = 30, color = "black"),
    axis.title.x = eb(),
    axis.text = et(size = 25, color = "black"),
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5
    ) 
  ) +
  ylab("Alpha Diversity\n(Shannon Index)\n") +
  scale_fill_manual(values = rev(pirate_colors)) +
  scale_color_manual(values = rev(pirate_colors)) +
  scale_y_continuous(breaks = seq(0, 5, 1))

# Figure 1B: Top Left Panel
gg_alpha_shannon + ggtitle("Figure 1B: Left Panel")

pdf("./Results/Pirate_Shannon_train.pdf",
  height = 6,
  width = 7
)
gg_alpha_shannon
invisible(dev.off())

#### Species Richness START ####
mat_richness <- mat_filt
row.names(mat_richness) <- mat_richness$shotgunSeq_id
mat_richness <- mat_richness %>% select(-shotgunSeq_id)

mat_richness_t <- mat_richness %>% t()

alpha_richness <- vegan::specnumber(mat_richness) %>%
  as.data.frame()
colnames(alpha_richness)[1] <- "Richness"
alpha_richness$shotgunSeq_id <- row.names(alpha_richness)

# Obtain values for mean alpha diversity for Survivor and Non-Survivor
alpha_richness %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  group_by(thirtyday_mortality_overall) %>%
  summarise(mean = mean(Richness))

# Obtain stats for species richness
alpha_richness_stats <-
  alpha_richness %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  mutate(
    thirtyday_mortality_overall = as.factor(thirtyday_mortality_overall),
    thirtyday_mortality_overall = factor(
      thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  ) %>%
  rstatix::wilcox_test(Richness ~ thirtyday_mortality_overall)

set.seed(456)
gg_alpha_richness <- alpha_richness %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  mutate(
    thirtyday_mortality_overall = as.factor(thirtyday_mortality_overall),
    thirtyday_mortality_overall = factor(
      thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  ) %>%
  ggplot(
    .,
    aes(
      x = thirtyday_mortality_overall,
      y = Richness,
      colour = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  geom_pirate(
    cis_params = list(fill = "white", alpha = 0.5),
    bars_params = list(alpha = 0.65),
    lines_params = list(size = 0.5),
    points_params = list(fill = "black", size = 3.5),
    jitter_width = 0.75,
    cis = TRUE,
    violins = FALSE
  ) +
  annotate(
    "text",
    x = 1.5,
    y = 175,
    label = paste0(
      "Wilcoxon, W = ",
      alpha_richness_stats$statistic,
      ", p = ",
      alpha_richness_stats$p
    ),
    size = 8
  ) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title.y = et(size = 30, color = "black"),
    axis.title.x = eb(),
    axis.text = et(size = 25, color = "black"),
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5
      )
  ) +
  ylab("Alpha Diversity\n(Species Richness)\n") +
  scale_fill_manual(values = pirate_colors) +
  scale_color_manual(values = pirate_colors) +
  scale_y_continuous(breaks = seq(0, 175, 25))


# Figure 1B: Middle Panel
gg_alpha_richness + ggtitle("Figure 1B: Middle Panel")

pdf("./Results/Pirate_Richness_train.pdf",
  height = 6,
  width = 7
)
gg_alpha_richness
invisible(dev.off())

#### Species Evenness START ####
mat_evenness <- mat_filt
row.names(mat_evenness) <- mat_evenness$shotgunSeq_id
mat_evenness <- mat_evenness %>% select(-shotgunSeq_id)
mat_evenness_t <- mat_evenness %>% t()
h <- vegan::diversity(mat_evenness)
s <- vegan::specnumber(mat_filt)
alpha_evenness <- h / log(s)
alpha_evenness <- as.data.frame(alpha_evenness)
colnames(alpha_evenness)[1] <- "Evenness"
alpha_evenness$shotgunSeq_id <- row.names(alpha_evenness)

# Obtain values for mean alph for Survivor and Non-Survivor
alpha_evenness %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  group_by(thirtyday_mortality_overall) %>%
  summarise(mean = mean(Evenness))

# Obtain stats for alpha diversity
alpha_evenness_stats <-
  alpha_evenness %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  mutate(
    thirtyday_mortality_overall = as.factor(thirtyday_mortality_overall),
    thirtyday_mortality_overall = factor(
      thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  ) %>%
  rstatix::wilcox_test(Evenness ~ thirtyday_mortality_overall)

set.seed(456)
gg_alpha_evenness <- alpha_evenness %>%
  left_join(micu_new_nocovid_oc %>% select(shotgunSeq_id, thirtyday_mortality_overall)) %>%
  mutate(
    thirtyday_mortality_overall = as.factor(thirtyday_mortality_overall),
    thirtyday_mortality_overall = factor(
      thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  ) %>%
  ggplot(
    .,
    aes(
      x = thirtyday_mortality_overall,
      y = Evenness,
      colour = thirtyday_mortality_overall,
      fill = thirtyday_mortality_overall
    )
  ) +
  geom_pirate(
    cis_params = list(fill = "white", alpha = 0.5),
    bars_params = list(alpha = 0.65),
    lines_params = list(size = 0.5),
    points_params = list(fill = "black", size = 3.5),
    jitter_width = 0.75,
    cis = TRUE,
    violins = FALSE
  ) +
  annotate(
    "text",
    x = 1.5,
    y = 1,
    label = paste0(
      "Wilcoxon, W = ",
      alpha_evenness_stats$statistic,
      ", p = ",
      alpha_evenness_stats$p
    ),
    size = 8
  ) +
  theme_bw() +
  theme(
    panel.grid = eb(),
    axis.title.y = et(size = 30, color = "black"),
    axis.title.x = eb(),
    axis.text = et(size = 25, color = "black"),
    plot.margin = margin(
      # Top margin
      t = 5,
      # Right margin
      r = 5,
      # Bottom margin
      b = 5,
      # Left margin
      l = 5
      ) 
  ) +
  ylab("Alpha Diversity\n(Species Evenness)\n") +
  scale_fill_manual(values = pirate_colors) +
  scale_color_manual(values = pirate_colors) +
  scale_y_continuous(breaks = seq(0, 1, 0.1))


# Figure 1B: Right Panel
gg_alpha_evenness + ggtitle("Figure 1B: Right Panel")

pdf("./Results/Pirate_Evenness_train.pdf",
  height = 6,
  width = 7
)
gg_alpha_evenness
invisible(dev.off())
```

#### Delta SOFA Score Correlation with MD Score
```{r, delta-sofa-score}
delta_stool_sofa <- micu_new_nocovid_oc %>%
  select(metabolomicsID, dSOFA_admission, dSOFA_stool) %>%
  left_join(km_nocovid_final %>% select(metabolomicsID, thirtyday_mortality_overall, md_score)) %>%
  mutate(
    md_score = as.numeric(md_score),
    dSOFA_admission = as.numeric(dSOFA_admission)
  ) %>%
  drop_na(thirtyday_mortality_overall)

# Delta SOFA Stool
ggscatter(
  delta_stool_sofa,
  y = "dSOFA_stool",
  x = "md_score",
  size = 3,
  alpha = 0.2,
  palette = "jco",
  add = "reg.line"
) +
  stat_cor(
    method = "spearman"
  )

ggsave(
  filename = "./Results/delta_SOFA_stool_MDS_train.pdf",
  height = 8,
  width = 8,
  units = "in"
)

# Delta SOFA Stool
ggscatter(
  delta_stool_sofa,
  y = "dSOFA_admission",
  x = "md_score",
  size = 3,
  alpha = 0.2,
  palette = "jco",
  add = "reg.line"
) +
  stat_cor(
    method = "spearman"
  )

ggsave(
  filename = "./Results/delta_SOFA_admission_MDS_train.pdf",
  height = 8,
  width = 8,
  units = "in"
)
```

#### Diversity, MD Score Patient List
```{r, shannon-mds-list}
shannon_mmp_list <- micu_new_nocovid_oc %>%
  select(shotgunSeq_id, metabolomicsID) %>%
  right_join(km_nocovid_final %>% select(metabolomicsID, thirtyday_mortality_overall, md_score)) %>%
  left_join(cutpoints_results_var_slct_shannon) %>%
  mutate(
    High_Shannon = ifelse(
      Shannon >= coordinates_shannon$threshold,
      "High Diversity",
      "Low Diversity"
    ),
    HS_LMDS = ifelse(
      Shannon >= coordinates_shannon$threshold &
        md_score < coordinates_mds$threshold,
      1,
      0
    ),
    HS_HMDS = ifelse(
      Shannon >= coordinates_shannon$threshold &
        md_score >= coordinates_mds$threshold,
      1,
      0
    ),
    LS_LMDS = ifelse(
      Shannon < coordinates_shannon$threshold &
        md_score < coordinates_mds$threshold,
      1,
      0
    ),
    LS_HMDS = ifelse(
      Shannon < coordinates_shannon$threshold &
        md_score >= coordinates_mds$threshold,
      1,
      0
    )
  )

write.csv(shannon_mmp_list,
  "./Results/shannon_mds_list_train.csv",
  row.names = FALSE
)

# Summary
shannon_mmp_list %>%
  select(HS_LMDS:LS_HMDS) %>%
  colSums() %>%
  as.data.frame() %>%
  rownames_to_column(var = "measure") %>%
  dplyr::rename(count = ".") %>%
  mutate(
    total = sum(count),
    percent = (count / total) * 100
  ) %>%
  write.csv(., "./Results/shannon_mds_summary_train.csv", row.names = FALSE)
```

#### First Sample Distribution
```{r, first-sample-distribution}
# Make dataframe of first samples and their day_collected
first_samp_dist <- first_samp_list_anon %>%
  right_join(micu_new_nocovid_oc %>%
    select(unique_id, thirtyday_mortality_overall)) %>%
  select(thirtyday_mortality_overall, day_collected)

# Run Kolmogorov-Smirnov test to compare distributions of day_collected between Survivor and Non-Survivor
first_samp_test <-
  ks.test(
    first_samp_dist %>% filter(thirtyday_mortality_overall == "Survivor") %>% pull(day_collected),
    first_samp_dist %>% filter(thirtyday_mortality_overall != "Survivor") %>% pull(day_collected)
  )

first_samp_test
# D = 0.14 and p = 0.24, therefore the samples from both "Survivor" and "Non-Survivor" come from the same distribution

first_samp_dist %>%
  ggpubr::gghistogram(
    x = "day_collected",
    fill = "thirtyday_mortality_overall",
    color = "thirtyday_mortality_overall",
    binwidth = 1,
    palette = c(
      ggsci::pal_lancet(palette = "lanonc")(2)[1],
      ggsci::pal_lancet(palette = "lanonc")(2)[2]
    ),
    alpha = 0.3
  ) +
  annotate(
    "text",
    x = 15,
    y = 15,
    label = paste0(
      "Kolmogorov-Smirnov; D(195) = ",
      round(first_samp_test[1][[1]], 3),
      " p = ",
      round(first_samp_test[2][[1]], 3)
    )
  ) +
  guides(
    fill = guide_legend("Outcome"),
    color = guide_legend("Outcome")
  ) +
  ylab("Count\n") +
  xlab("\nDay of First Sample Collection")

ggsave(
  "./Results/first_samp_distribution_train.pdf",
  height = 6,
  width = 8,
  units = "in"
)
```

#### Microbiome Metabolomic Profile 
```{r, microbiome-metabolomic-profile}
# MMP Score
mmp_df <- cutpoints_df %>% 
  filter(compound %in% c("deoxycholic acid", 
                         "isodeoxycholic acid", 
                         "lithocholic acid", 
                         "desaminotyrosine")) %>% 
  mutate(cutpoint_prediction = case_when(
      compound == "deoxycholic acid" &
        mvalue__mM >= (89.92/1000) ~ 0,
      compound == "isodeoxycholic acid" &
        mvalue__mM >= (0.97/1000) ~ 0,
      compound == "lithocholic acid" &
        mvalue__mM >= (258.25/1000) ~ 0,
      compound == "desaminotyrosine" &
        mvalue__mM >= (21.31/1000) ~ 0,
      TRUE ~ 1
    )) %>% 
  group_by(metabolomicsID, thirtyday_mortality_overall) %>%
  summarize(mmp_score = sum(cutpoint_prediction)) %>% 
  ungroup() %>% 
  mutate(grouped_mmp_score = ifelse(mmp_score >= 2, "High MMP", "Low MMP"))

  
# Boxplot of MD Score
mmp_chis <-
  stats::chisq.test(
    mmp_df$thirtyday_mortality_overall,
    mmp_df$grouped_mmp_score
  )

mmp_violin <-
  ggviolin(
    mmp_df,
    x = "thirtyday_mortality_overall",
    y = "mmp_score",
    fill = "thirtyday_mortality_overall",
    palette = "lancet",
    add = c("dotplot"),
    add.params = list(binwidth = 0.05)
  ) +
  annotate(
    "text",
    x = 1.5,
    y = 12,
    label = paste0(
      "Chisq",
      "(",
      round(mmp_chis$statistic, 3),
      "),",
      " p =",
      scales::scientific(mmp_chis$p.value)
    )
  ) +
  annotate(
    "segment",
    x = 1,
    xend = 2,
    y = 11.35,
    yend = 11.35
  ) +
  annotate(
    "segment",
    x = 1,
    xend = 1,
    y = 11.25,
    yend = 11.35
  ) +
  annotate(
    "segment",
    x = 2,
    xend = 2,
    y = 11.25,
    yend = 11.35
  ) +
  ylab("Microbiome Metabolomic Profile\n") +
  xlab("") +
  guides(fill = guide_legend("30 Day Mortality"))

mmp_violin

ggsave(
  plot = mmp_violin,
  filename = "./Results/MMP_Violin_train.pdf",
  height = 6,
  width = 8
)


# gg_mmp_chi <- gginference::ggchisqtest(mmp_chis, colaccept = "green3", colreject = "red3") # It is highly unlikely that our test statistic would be observed if there were no association between survival outcome and the md score
# gg_mmp_chi

# Confusion matrix for MMP Score: covid score

mmp_df2 <- mmp_df %>% 
  mutate(prediction = ifelse(grouped_mmp_score == "Low MMP", "Survivor", "Non-Survivor"))

caret::confusionMatrix(table(
  factor(mmp_df2$prediction,
         levels = c("Survivor", "Non-Survivor")),
  factor(
    mmp_df2$thirtyday_mortality_overall,
    levels = c("Survivor", "Non-Survivor")
  )
))

# Confusion Matrix and Statistics
# 
#               
#                Survivor Non-Survivor
#   Survivor           42            8
#   Non-Survivor       60           37
#                                           
#                Accuracy : 0.5374          
#                  95% CI : (0.4534, 0.6199)
            # Sensitivity : 0.4118          
            # Specificity : 0.8222  

# Export confusion matrix
bind_rows(
  as.data.frame(as.table(caret::confusionMatrix(
  table(
    factor(mmp_df2$prediction,
           levels = c("Survivor", "Non-Survivor")),
    factor(
      mmp_df2$thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  )
))) %>% dplyr::rename(Prediction = Var1,
                      Actual = Var2),

  as.data.frame(as.matrix(caret::confusionMatrix(
  table(
    factor(mmp_df2$prediction,
           levels = c("Survivor", "Non-Survivor")),
    factor(
      mmp_df2$thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  )
),
what = "overall")),

as.data.frame(as.matrix(caret::confusionMatrix(
  table(
    factor(mmp_df2$prediction,
           levels = c("Survivor", "Non-Survivor")),
    factor(
      mmp_df2$thirtyday_mortality_overall,
      levels = c("Survivor", "Non-Survivor")
    )
  )
),
what = "classes"))) %>% write.csv("./Results/MMP_Confusion_Matrix_Data.csv")
```

#### Kaplan-Meier Survival Analysis
```{r}
km_mmp <- micu_new_nocovid_oc %>%
  select(
    unique_id,
    sampleid,
    metabolomicsID,
    days_until_death_overall,
    censoring_thirtyday_mortality_overall,
    thirtyday_mortality_overall
  ) %>%
  ungroup() %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>% # Non-Survivor is 1
  left_join(mmp_df)
  

# KM Curves: MD Score
set.seed(123)
surv_object_mmp <-
  Surv(
    time = km_mmp$surv_days,
    event = km_mmp$thirtyday_mortality_overall_class
  )

fit_mmp <- survfit(surv_object ~ grouped_mmp_score, data = km_mmp)

ggs_mmp <- ggsurvplot(
  fit_mmp,
  data = km_mmp,
  size = 1,
  palette = c("#C45258", "#107b1c"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  legend.labs = c("High MMP Score", "Low MMP Score")
)

# Change table axis labels
ggs_mmp$table <-
  ggs_mmp$table + labs(x = NULL, y = NULL) + theme(plot.title = eb()) # risk table

ggs_mmp

pdf(
  "./Results/kaplan_meier_MMP_30_Day_Mortality_train.pdf",
  height = 4,
  width = 6,
  onefile = FALSE
)
ggs_mmp
invisible(dev.off())
```

#### Cox Proportional Hazards Regression Analysis
```{r}
# Variables labels
cox_df_mmp <- tableone_nocovid_df_filt %>%
  labelled::remove_labels() %>%
  janitor::clean_names() %>%
  mutate(
    race_factor = as.character(race_factor),
    race_factor = ifelse(
      race_factor %in% c("Asian", "More than one race", "White, Hispanic"),
      "Other",
      race_factor
    )
  ) %>%
  left_join(
    micu_nocovid_first_samps_omics_light %>%
      group_by(metabolomicsID) %>%
      slice(1) %>%
      select(unique_id, metabolomicsID)
  ) %>%
  left_join(km_mmp %>% select(metabolomicsID, mmp_score)) %>%
  right_join(
    micu_new_nocovid_oc %>% select(
      unique_id,
      days_until_death_overall,
      censoring_thirtyday_mortality_overall,
      thirtyday_mortality_overall
    )
  ) %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)) %>% 
  dplyr::rename(`Charlson Comorbidity Index` = cci_total_sc) %>%
  mutate(diet = ifelse(diet == "1", "Diet", "NPO")) %>% 
  dplyr::rename(
    `Sex` = "sex_factor",
    `Age` = "age",
    `Acute respiratory distress syndrome` = "ards_factor",
    `Sepsis` = "sepsis_factor",
    `SOFA Score` = "sofa_score_total",
    `Race` = "race_factor",
    `Time to stool sample` = "day_collected",
    `Diet` = "diet",
    `MMP` = "mmp_score",
  )


reset_gtsummary_theme()

coxauc_mmp <-
  coxph(
    Surv(cox_df_mmp$surv_days, cox_df_mmp$thirtyday_mortality_overall_class) ~
      `Sex` +
      `Age` +
      `Charlson Comorbidity Index` +
      `Acute respiratory distress syndrome` +
      `Sepsis` +
      `SOFA Score` +
      `Race` +
      `Time to stool sample` +
      `Diet` +
      `MMP`,
    data = cox_df_mmp
  ) %>%
  tbl_regression(
    exp = TRUE,
    pvalue_fun = function(x) {
      if_else(is.na(x), NA_character_, if_else(
        x < 0.001,
        format(x,
          digits = 3, scientific = TRUE
        ),
        format(round(x, 3),
          scientific = F
        )
      ))
    }
  ) %>%
  modify_footnote(everything() ~ NA, abbreviation = TRUE)


coxauc_mmp %>%
  gtsummary::modify_caption("**Cox Proportional Hazards Regression**")

# In case you get an error: "Error in s$close() : attempt to apply non-function", run this code below:
# f <- chromote::default_chromote_object() #get the f object
# f$close()

gt::gtsave(gtsummary::as_gt(coxauc_mmp), file = "./Results/cox_model_MMP_30_Day_Mortality_train.png")
```

## Validation Cohort
### Validation Cohort Analysis
#### MD Score
```{r}
# Cutpoint dataframe
cutpoints_df_vc <- metab_quant_imp_tot_mM %>%
  pivot_wider(
    id_cols = c(metabolomicsID),
    names_from = "compound",
    values_from = "mvalue__mM"
  ) %>%
  group_by(metabolomicsID) %>%
  pivot_longer(!c(metabolomicsID),
    names_to = "compound",
    values_to = "mvalue__mM"
  ) %>%
  right_join(micu_new_nocovid_vc %>% select(metabolomicsID, thirtyday_mortality_overall)) %>%
  group_by(compound) %>%
  mutate(n = length(compound)) %>%
  ungroup() %>%
  mutate(p = length(unique(metabolomicsID))) %>%
  mutate(
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>%
  drop_na(compound)

cutpoints_results_var_slct_vc <-
  cutpoints_df_vc %>%
  filter(
    compound %in% c(
      as.matrix(coef(cutpoint_best_ridge)) %>%
        as.data.frame() %>%
        rownames_to_column(var = "compound") %>%
        filter(compound != "(Intercept)") %>%
        arrange(desc(abs(s0))) %>%
        dplyr::slice(1:optimal_components) %>%
        pull(compound)
    )
  ) %>%
  left_join(
    cutpoints_unnest %>%
      dplyr::rename(compound = subgroup) %>%
      select(compound, direction, optimal_cutpoint)
  ) %>%
  mutate(
    cutpoint_prediction = case_when(
      direction == "<=" & mvalue__mM <= optimal_cutpoint ~ 1,
      direction == "<=" & mvalue__mM > optimal_cutpoint ~ 0,
      direction == ">=" & mvalue__mM >= optimal_cutpoint ~ 1,
      direction == ">=" & mvalue__mM < optimal_cutpoint ~ 0
    )
  ) %>%
  group_by(metabolomicsID, thirtyday_mortality_overall) %>%
  summarize(md_score = sum(cutpoint_prediction)) %>%
  mutate(
    grouped_md_score = ifelse(
      md_score >= coordinates_mds$threshold,
      "High Score",
      "Low Score"
    ),
    prediction = ifelse(grouped_md_score == "Low Score", "Survivor", "Non-Survivor")
  )


cutpoints_results_var_slct_vc <- cutpoints_results_var_slct_vc %>% 
  mutate(thirtyday_mortality_overall_class = if_else(thirtyday_mortality_overall == "Survivor", 0, 1))

pROC::roc(cutpoints_results_var_slct_vc$thirtyday_mortality_overall_class,
          cutpoints_results_var_slct_vc$md_score)

# AUC: 0.6833

vc_cm <- caret::confusionMatrix(table(
  factor(
    cutpoints_results_var_slct_vc$prediction,
    levels = c("Survivor","Non-Survivor")
  ),
  factor(
    cutpoints_results_var_slct_vc$thirtyday_mortality_overall,
    levels = c("Survivor","Non-Survivor")
  )
))

 #                 Survivor Non-Survivor
 #  Survivor           28            8
 #  Non-Survivor        6            7
 #                                          
 #               Accuracy : 0.7143          
 #                 95% CI : (0.5674, 0.8342)
 #    No Information Rate : 0.6939          
 #    P-Value [Acc > NIR] : 0.4464          
 #                                          
 #                  Kappa : 0.3014          
 #                                          
 # Mcnemar's Test P-Value : 0.7893          
 #                                          
 #            Sensitivity : 0.8235          
 #            Specificity : 0.4667  

print(vc_cm)

library(epiR)

data <- as.table(matrix(c(28,8,6,7), nrow = 2, byrow = TRUE))
rval <- epi.tests(data, conf.level = 0.95)

print(rval)

```

#### Kaplan-Meier Survival Analysis
```{r}
km_nocovid_vc <- micu_new_nocovid_vc %>%
  select(
    unique_id,
    sampleid,
    metabolomicsID,
    days_until_death_overall,
    censoring_thirtyday_mortality_overall,
    thirtyday_mortality_overall
  ) %>%
  ungroup() %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>% # Non-Survivor is 1
  left_join(cutpoints_results_var_slct_vc) 

# KM Curves: MD Score
set.seed(123)
surv_object_vc <-
  Surv(
    time = km_nocovid_vc$surv_days,
    event = km_nocovid_vc$thirtyday_mortality_overall_class
  )

fit_vc <-
  survfit(surv_object_vc ~ grouped_md_score, data = km_nocovid_vc)

ggs_vc <- ggsurvplot(
  fit_vc,
  data = km_nocovid_vc,
  size = 1,
  palette = c("#C45258", "#2F4858"),
  xlab = "Days from Admission",
  conf.int = TRUE,
  pval = TRUE,
  risk.table = "abs_pct",
  legend = "bottom",
  risk.table.height = 0.4,
  risk.table.y.text.col = TRUE,
  tables.y.text = FALSE,
  risk.table.fontsize = 2.8,
  pval.size = 3.5,
  ggtheme = theme_test() + theme(
    panel.grid.major = el(linewidth = 0.5, color = "gray90"),
    axis.text.y = et(color = "black", size = 10),
    axis.title.y = et(color = "black")
  ),
  legend.labs = c("High MD Score", "Low MD Score")
)

# Change table axis labels
ggs_vc$table <-
  ggs_vc$table + labs(x = NULL, y = NULL) + theme(plot.title = eb()) # risk table

ggs_vc

pdf(
  "./Results/kaplan_meier_roc_loop_30_Day_Mortality_validation.pdf",
  height = 4,
  width = 6,
  onefile = FALSE
)
ggs_vc
invisible(dev.off())

# Boxplot of MD Score
mds_chis_vc <-
  stats::chisq.test(
    km_nocovid_vc$thirtyday_mortality_overall,
    km_nocovid_vc$md_score
  )

md_violin_vc <-
  ggviolin(
    km_nocovid_vc,
    x = "thirtyday_mortality_overall",
    y = "md_score",
    fill = "thirtyday_mortality_overall",
    palette = "lancet",
    add = c("dotplot"),
    add.params = list(binwidth = 0.05)
  ) +
  annotate(
    "text",
    x = 1.5,
    y = 12,
    label = paste0(
      "Chisq",
      "(",
      round(mds_chis_vc$statistic, 3),
      "),",
      " p =",
      scales::scientific(mds_chis_vc$p.value)
    )
  ) +
  annotate(
    "segment",
    x = 1,
    xend = 2,
    y = 11.35,
    yend = 11.35
  ) +
  annotate(
    "segment",
    x = 1,
    xend = 1,
    y = 11.25,
    yend = 11.35
  ) +
  annotate(
    "segment",
    x = 2,
    xend = 2,
    y = 11.25,
    yend = 11.35
  ) +
  ylab("Metabolic Dysbiosis Score\n") +
  xlab("") +
  guides(fill = guide_legend("30 Day Mortality"))

md_violin_vc

ggsave(
  plot = md_violin_vc,
  filename = "./Results/MDS_Violin_validation.pdf",
  height = 6,
  width = 8
)

km_nocovid_vc |> 
  group_by(thirtyday_mortality_overall) |> 
  summarise(mean = mean(md_score),
            sd = sd(md_score))

stats::wilcox.test(
    km_nocovid_vc$md_score ~ km_nocovid_vc$thirtyday_mortality_overall
  )

# gg_mds_chi_vc <- gginference::ggchisqtest(mds_chis_vc, colaccept = "green3", colreject = "red3") # It is highly unlikely that our test statistic would be observed if there were no association between survival outcome and the md score
# gg_mds_chi_vc
```

#### Cox Proportional Hazards Regression Analysis
```{r}
# Validation cohort 
cri_rxmar_abx_long_vc <- readRDS("./Data/cri_rxmar_abx_long_vc.rds")

tableone_nocovid_df_vc <-
  micu_new_nocovid_vc %>%
  left_join(cri_rxmar_abx_long_vc, by = "unique_id") %>% 
  mutate(across(Cephalosporins:Quinolones, ~ str_to_title(.))) %>% 
  mutate(across(Cephalosporins:Quinolones, ~ replace_na(., "Unchecked"))) %>%
  mutate(across(Cephalosporins:Quinolones, ~ as.factor(.))) %>% 
  mutate(across(Cephalosporins:Quinolones, ~ factor(., levels = c("Unchecked", "Checked")))) %>% 
  mutate(across(Hypertension:Tuberculosis, ~ factor(., levels = c("Unchecked", "Checked")))) %>%
  mutate(across(Acute.respiratory.distress.syndrome:Newly.diagnosed.solid.malignancy, ~ factor(., levels = c("Unchecked", "Checked")))) %>%
  mutate(across(Myocardial.infract:AIDS, ~ factor(., levels = c("Unchecked", "Checked")))) %>%
  select(
    age,
    sex.factor,
    bmi,
    race.factor,
    cci_total_sc,
    thirtyday_mortality_overall,
    primary_dx.factor,
    ards.factor,
    sepsis.factor,
    admit_from.factor,
    COVID_upon_admission,
    sofa_score_total,
    ap2_total_score,
    reason_for_intubation.factor,
    reintub_1.factor,
    reintub_2.factor,
    total_ventilator_days,
    icu_los_total,
    hospital_los,
    day_collected,
    Hypertension:`Neuromuscular.disorder`,
    `Peptic.ulcer.disease`,
    `Thyroid.disease`:Tuberculosis,
    `Bacterial.pneumonia`:`Newly.diagnosed.solid.malignancy`,
    `Myocardial.infract`:`AIDS`,
    Penicillins,
    Cephalosporins,
    Carbapenems,
    Vancomycin,
    Metronidazole,
    Macrolides,
    Quinolones,
    other,
    Clindamycin,
    Aminoglycosides,
    Doxycycline,
    `Trimethoprim-Sulfamethoxazole`,
    Rifaximin,
    `diet`,
    dSOFA_admission, dSOFA_stool
  ) %>%
  janitor::clean_names() %>%
  select(-c(
    hypertension:tuberculosis,
    reason_for_intubation_factor:hospital_los
  )) %>%
  replace_na(list(reason_for_intubation_factor = "Not intubated")) %>%
  droplevels()

tableone_nocovid_vc <- CreateTableOne(
  data = tableone_nocovid_df_vc,
  strata = "thirtyday_mortality_overall",
  includeNA = TRUE
)

summary(tableone_nocovid_vc)

# Print tableone
tableone_nocovid_print_vc <-
  print(
    tableone_nocovid_vc,
    nonnormal = TRUE,
    formatOptions = list(big.mark = ",")
  )

# Save CSV
write.csv(
  tableone_nocovid_print_vc,
  "./Results/Table_One_30_Days_Mortality_validation.csv",
  row.names = TRUE
)

# Save table for paper
tableone_nocovid_print_vc_clean <-
tableone_nocovid_print_vc %>%
  as.data.frame() %>%
  rownames_to_column(var = "variable") %>% #distinct(variable)
  mutate(
    variable = dplyr::recode(
      variable,
            n = "Number of Patients",
      `age (median [IQR])` = "Age (median [IQR])",
      `sex_factor = Male (%)` = "Male (%)",
      `bmi (median [IQR])` = "Body Mass Index (median [IQR])",
      `race_factor (%)` = "Race (%)",
      `African American` = " African American",
      `Asian` = "Asian",
      `More than one race` = "More than one race",
      `White, Hispanic` = "White, Hispanic",
      `White, non-Hispanic` = "White, Non-Hispanic",
      `Other` = "Unknown Race",
      `cci_total_sc (median [IQR])` = "Charlson Comorbidity Index (median [IQR])",
      `primary_dx_factor (%)` = "Primary admission diagnosis (%)",
      `Acute (on chronic) liver failure` = "Acute chronic liver failure",
      # `X...AMI.dysrhythmia` = "AMI dysrhytmia",
      # `X...CHF.cardiogenic.shock` = "CHF cardiogenic shock",
      `CNS pathology` = "CNS pathology",
      `GI hemorrhage` = "GI hemorrhage",
      # `X...Metabolic` = "Metabolic",
      # `X...Other` = "Other Primary diagnosis",
      `Post-operative observation` = "Post-operative observation",
      `Respiratory failure, AHRF` = "Respiratory failure (AHRF)",
      `Respiratory failure, airway compromise` = "Respiratory failure, airway compromise",
      `Respiratory failure, ventilatory` = "Respiratory failure, ventilatory",
      `Sepsis (+/- septic shock)` = "Sepsis, septic shock",
      `ards_factor = Yes (%)` = "Acute respiratory distress syndrome (%)",
      `sepsis_factor = Sepsis (%)` = "Sepsis (%)",
      `admit_from_factor (%)` = "Admitted from (%)",
      `Cardiology` = "Cardiology",
      `ED` = "Emergency Department",
      `General Medicine` = "General Medicine",
      `Liver` = "Liver",
      # `Neurology` = "Nuerology",
      `Oncology` = "Oncology",
      `OSH` = "Outside Hospital",
      `Surgery` = "Surgery",
      # `X...NA.1` = "Unknown",
      `covid_upon_admission = No (%)` = "No Covid upon admission (%)",
      `sofa_score_total (median [IQR])` = "SOFA Score (median [IQR])",
      `ap2_total_score (median [IQR])` = "APACHE II Score (median [IQR])",
      `day_collected (median [IQR])` = "Day From Admission Stool Sample Collected (median [IQR])",
      `bacterial_pneumonia = Checked (%)` = "Bacterial Pneumonia (%)",
      `fungal_pneumonia = Checked (%)` = "Fungal Pneumonia (%)",
      `viral_pneumonia = Checked (%)` = "Viral Pneumonia (%)",
      `chronic_obstructive_pulmonary_disease_copd_1 = Checked (%)` = "Chronic Obstructive Pulmonary Disease (COPD) (%)",
      `asthma_exacerbation = Unchecked (%)` = "Asthma exacerbation (%)",
      `lung_lobar_collapse = Checked (%)` = "Lung/lobar collapse (%)",
      `pulmonary_embolism = Checked (%)` = "Pulmonary embolism (%)",
      `hemoptysis = Unchecked (%)` = "Hemoptysis (%)",
      `pancreatitis = Unchecked (%)` = "Pancreatitis (%)",
      `infection_genitourinary_system = Checked (%)` = "Infection, genitourinary system (%)",
      `infection_intra_abdominal = Checked (%)` = "Infection, Intra-abdominal (%)",
      `infection_soft_tissue = Checked (%)` = "Infection, soft tissue (%)",
      `infection_cns = Checked (%)` = "Infection, CNS (%)",
      `hepatic_failure_acute_fullminant = Unchecked (%)` = "Hepatic failure, acute fullminant (%)",
      `hepatic_failure_acute_on_chronic = Checked (%)` = "Hepatic failure, acute on chronic (%)",
      `diabetic_ketoacidosis = Checked (%)` = "Diabetic ketoacidosis (%)",
      `acute_leukemia = Unchecked (%)` = "Acute leukemia (%)",
      `cerebral_vascular_accident_1 = Checked (%)` = "Cerebreal vascular accident (%)",
      `acute_myocardial_infarction_nstemi_stemi = Unchecked (%)` = "Acute myocardial infarction (NSTEMI/STEMI) (%)",
      `diffuse_alveolar_hemorrhage = Unchecked (%)` = "Diffuse alveolar hemorrhage (%)",
      `decompensated_heart_failure_pulmonary_oedema = Checked (%)` = "Decompensated heart failure/Pulmonary oedema (%)",
      `pleural_effusion = Checked (%)` = "Pleural effusion (%)",
      `interstitial_lung_disease_exacerbation = Checked (%)` = "Interstitial lung disease exacerbation (%)",
      `organizing_pneumonia = Unchecked (%)` = "Organizing pneumonia (%)",
      `acute_eosinophilic_pneumoniae = Unchecked (%)` = "Acute eosinophilic pneumoniae (%)",
      `other = Checked (%)` = "Other (%)",
      `angioedema = Checked (%)` = "Angioedema (%)",
      `acute_renal_failure = Checked (%)` = "Acute renal failure (%)",
      `altered_mental_status = Checked (%)` = "Altered mental status (%)",
      `hypertensive_urgency = Checked (%)` = "Hypertensive urgency (%)",
      `hypertensive_emergency = Checked (%)` = "Hypertensive emergency (%)",
      `endocarditis = Checked (%)` = "Endocarditis (%)",
      `bacteremia = Checked (%)` = "Bacteremia (%)",
      `gastrointestinal_bleeding = Unchecked (%)` = "Gastrointestinal bleeding (%)",
      `hemorrhagic_shock = Unchecked (%)` = "Hemorrhagic shock (%)",
      `aspiration = Checked (%)` = "Aspiration (%)",
      `central_line_associated_blood_steam_infection = Unchecked (%)` = "Central line associated blood steam infection (%)",
      `prosthetic_joint_infection = Unchecked (%)` = "Prosthetic joint infection (%)",
      `new_onset_atrial_fibrillation = Unchecked (%)` = "New onset atrial fibrillation (%)",
      `newly_diagnosed_solid_malignancy = Checked (%)` = "Newly diagnosed solid malignancy (%)",
      `myocardial_infract = Checked (%)` = "Myocardial infract (%)",
      `congestive_heart_failure = Checked (%)` = "Congestive heart failure (%)",
      `peripheral_vascular_disease_cci = Checked (%)` = "Peripheral vascular disease (%)",
      `cerebrovascular_disease = Checked (%)` = "Cerebrovascular disease (%)",
      `dementia = Unchecked (%)` = "Dementia (%)",
      `chronic_pulmonary_disease = Checked (%)` = "Chronic pulmonary disease (%)",
      `connective_tissue_disease_1 = Checked (%)` = "Connective tissue disease (%)",
      `ulcer_disease = Checked (%)` = "Ulcer disease (%)",
      `mild_liver_disease = Checked (%)` = "Mild liver disease (%)",
      `diabetes_without_complications = Checked (%)` = "Diabetes (without complications) (%)",
      `diabetes_with_end_organ_damage = Checked (%)` = "Diabetes (with end organ damage) (%)",
      `hemiplegia = Checked (%)` = "Hemiplegia (%)",
      `moderate_or_severe_renal_disease = Checked (%)` = "Moderate or severe renal disease (%)",
      `solid_tumor_non_metastatic = Checked (%)` = "Solid tumor (non-metastatic) (%)",
      `leukemia = Checked (%)` = "Leukemia (%)",
      `lymhoma = Checked (%)` = "Lymphoma (%)",
      `moderate_or_severe_liver_disease = Checked (%)` = "Moderate or severe liver disease (%)",
      `metastatic_solid_tumor = Checked (%)` = "Solid tumor (metastatic) (%)",
      `aids = Checked (%)` = "AIDS (%)",
      `penicillins = Checked (%)` = "Penicillins (%)",
      `cephalosporins = Checked (%)` = "Cephalosporins (%)",
      `carbapenems = Checked (%)` = "Carbapenems (%)",
      `vancomycin = Checked (%)` = "Vancomycin (%)",
      `metronidazole = Checked (%)` = "Metronidazole (%)",
      `macrolides = Checked (%)` = "Macrolides (%)",
      `quinolones = Checked (%)` = "Quinolones (%)",
      `other_2 = Checked (%)` = "Other Antiobiotics (%)",
      `clindamycin = Checked (%)` = "Clindamycin (%)",
      `aminoglycosides = Checked (%)` = "Aminoglycosides (%)",
      `doxycycline = Checked (%)` = "Doxycycline (%)",
      `trimethoprim_sulfamethoxazole = Checked (%)` = "Trimethoprim-Sulfamethoxazole (%)",
      `rifaximin = Checked (%)` = "Rifaximin (%)",
      `diet...npo....` = "Diet (nothing by mouth) (%)",
      `d_sofa_admission..median..IQR..` = "SOFA from admission (median [IQR])",
      `d_sofa_stool..median..IQR..` = "SOFA from Stool Sample (median [IQR])"
    )
  ) %>% 
  column_to_rownames(var = "variable")

# Export to csv to then load in as a dataframe
write.csv(
  tableone_nocovid_print_vc_clean,
  "./Results/Table_One_30_Days_Mortality_validation_clean.csv",
  row.names = TRUE
)

# Import csv as dataframe
tableone_nocovid_csv_vc <-
  read.csv("./Results/Table_One_30_Days_Mortality_validation.csv",
    stringsAsFactors = FALSE
  )

# Filter for only p-values <= 0.3 to then include in multi-variable model
tableone_pval_filt_vc <- tableone_nocovid_csv_vc %>%
  dplyr::rename(variable = X) %>%
  mutate(
    p = ifelse(p == "<0.001", 0.001, p),
    p = as.numeric(p)
  ) %>%
  # dplyr::slice(2:5, 11, 22:23, 33:35, 95:107) %>%
  filter(!grepl(variable, pattern = "^\\s"))

tableone_pval_filt_vars_vc <- tableone_pval_filt_vc %>%
  filter(variable != "n") %>%
  select(variable) %>%
  mutate(
    variable = as.character(variable),
    variable = gsub(
      x = variable,
      pattern = "\\s\\(median \\[IQR\\]\\)|\\s\\(%\\)| = Yes| = [Cc]hecked| = [Uu]nchecked| = Male| = npo| = Sepsis| = None",
      fixed = FALSE,
      replacement = ""
    )
  ) %>%
  filter(variable %!in% c("thirtyday_mortality_overall = Non-Survivor", "covid_upon_admission = No")) %>%
  pull(variable)

tableone_nocovid_df_filt_vc <-
  tableone_nocovid_df_vc[, tableone_pval_filt_vars_vc]

tableone_nocovid_df_filt_vc <- tableone_nocovid_df_filt_vc %>%
  bind_cols(
    micu_new_nocovid_vc %>% ungroup() %>%
      left_join(cri_rxmar_abx_long_vc, by = "unique_id") %>%
      mutate(across(
        Cephalosporins:Quinolones, ~ replace_na(., "unchecked")
      )) %>%
      mutate(across(
        Cephalosporins:Quinolones, ~ as.factor(.)
      )) %>%
      select(unique_id, thirtyday_mortality_overall)
  ) %>%
  relocate(unique_id, .before = NULL) %>%
  mutate_all(as.character) %>% 
  pivot_longer(
    !c(unique_id:day_collected, thirtyday_mortality_overall),
    names_to = "variable",
    values_to = "value"
  ) %>%
  mutate(
    value = as.character(value),
    value = ifelse(value %in% c("Checked", "checked", "diet"), 1, 0)
  ) %>% # diet = 1, npo = 0
  pivot_wider(names_from = "variable", values_from = "value") %>% 
    mutate(age = as.numeric(age),
         bmi = as.numeric(bmi),
         cci_total_sc = as.numeric(cci_total_sc),
         sofa_score_total = as.numeric(sofa_score_total),
         ap2_total_score = as.numeric(ap2_total_score),
         day_collected = as.numeric(day_collected)) %>% 
  mutate_if(is.character, as.factor)


# Variables labels
cox_df_vc <- tableone_nocovid_df_filt_vc %>%
  left_join(micu_new_nocovid_vc %>% distinct(unique_id, shotgunSeq_id, metabolomicsID)) %>%
  labelled::remove_labels() %>%
  janitor::clean_names() %>%
  mutate(
    race_factor = as.character(race_factor),
    race_factor = ifelse(
      race_factor %in% c("Asian", "More than one race", "White, Hispanic"),
      "Other",
      race_factor
    )
  ) %>%
  dplyr::rename(metabolomicsID = metabolomics_id) %>%
  left_join(cutpoints_results_var_slct_vc %>% select(metabolomicsID, md_score)) %>%
  mutate(grouped_md_score = ifelse(
    md_score >= coordinates_mds$threshold,
    "High Score",
    "Low Score"
  )) %>%
  right_join(
    micu_new_nocovid_vc %>% select(
      unique_id,
      days_until_death_overall,
      censoring_thirtyday_mortality_overall,
      thirtyday_mortality_overall
    )
  ) %>%
  mutate(
    surv_days = ifelse(
      is.na(days_until_death_overall) &
        thirtyday_mortality_overall == "Survivor",
      censoring_thirtyday_mortality_overall,
      days_until_death_overall
    ),
    surv_days = ifelse(
      is.na(surv_days) &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    surv_days = ifelse(
      surv_days > 30 &
        thirtyday_mortality_overall == "Survivor",
      30,
      surv_days
    ),
    thirtyday_mortality_overall_class = ifelse(thirtyday_mortality_overall == "Survivor", 0, 1)
  ) %>%
  group_by(metabolomicsID) %>%
  dplyr::slice(1) %>%
  dplyr::rename(`Charlson Comorbidity Index` = cci_total_sc) %>%
  mutate(diet = ifelse(diet == "1", "Diet", "NPO")) %>% 
  dplyr::rename(
    `Sex` = "sex_factor",
    `Age` = "age",
    `Acute respiratory distress syndrome` = "ards_factor",
    `Sepsis` = "sepsis_factor",
    `SOFA Score` = "sofa_score_total",
    `Race` = "race_factor",
    `Time to stool sample` = "day_collected",
    `Diet` = "diet",
    `MDS` = "md_score"
  )

reset_gtsummary_theme()

coxauc_vc <-
  coxph(
    Surv(
      cox_df_vc$surv_days,
      cox_df_vc$thirtyday_mortality_overall_class
    ) ~
      `Charlson Comorbidity Index` +
      `SOFA Score` +
      `MDS`,
    data = cox_df_vc
  ) %>%
  tbl_regression(
    exp = TRUE,
    pvalue_fun = function(x) {
      if_else(is.na(x), NA_character_, if_else(
        x < 0.001,
        format(x,
          digits = 3, scientific = TRUE
        ),
        format(round(x, 3),
          scientific = F
        )
      ))
    }
  ) %>%
  modify_footnote(everything() ~ NA, abbreviation = TRUE)


coxauc_vc %>%
  gtsummary::modify_caption("**Cox Proportional Hazards Regression**")

# In case you get an error: "Error in s$close() : attempt to apply non-function", run this code below:
# f <- chromote::default_chromote_object() #get the f object
# f$close()

gt::gtsave(gtsummary::as_gt(coxauc_vc), file = "cox/cox_model_SOFA_30_Day_Mortality_roc_loop_validation.png")
```

## Save Data Image
```{r save image, message=FALSE, warning=FALSE}
save.image(file = "./Data/MICU_Data_Anon.RData")
```
